<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Single Cell Shiny Application for Analysing Single Cell Transcriptomics Data • SCHNAPPs</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Single Cell Shiny Application for Analysing Single Cell Transcriptomics Data">
<meta property="og:description" content="Single Cell sHiny APPlication (SCHNAPPs) is a R/Shiny based application to interact, manipulate, explore, and analyze single cell RNA-seq experiments, including MARS-seq and others.
    The application is open and extendable. The main idea is to be able to run basic visualization and analyses, including normalization, clustering,...
    on a defined subset of cells/genes. Key is, that all computations are based only on the cells that have been selected. Input can be a 
    SingleCellExperiment object or CSV/TSV files of the count matrix.">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">SCHNAPPs</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.12.73</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/Readme.html">Readme</a>
    </li>
    <li>
      <a href="articles/pkdown/FAQs.html">FAQs</a>
    </li>
    <li>
      <a href="articles/pkdown/SCHNAPPs_usage.html">SCHNAPPs - Single Cell sHiNy APP(s)</a>
    </li>
    <li>
      <a href="articles/pkdown/SeuratWorkflow.html">Seurat workflow</a>
    </li>
    <li>
      <a href="articles/pkdown/scranWorkflow.html">Follow scran workflow to analyze single-cell RNA-seq data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/C3BI-pasteur-fr/UTechSCB-SCHNAPPs/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<p><img src="logo.png" align="right" width="120"></p>
<div class="section level1">
<div class="page-header"><h1 id="schnapps---single-cell-shiny-apps">SCHNAPPs - Single Cell sHiNy APP(s)<a class="anchor" aria-label="anchor" href="#schnapps---single-cell-shiny-apps"></a>
</h1></div>
<p><img src="https://travis-ci.com/C3BI-pasteur-fr/UTechSCB-SCHNAPPs.svg?branch=master"></p>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>Shiny app for the exploration and analysis of single cell RNAseq data as it comes from 10X or MARSseq technologies. It is currently being developed based on user requests of the Cytometry and Biomarkers UTechS at the Institut Pasteur, Paris. The goal is to enable the users of our platform to explore their data, select cells they would like to work with and then perform the final analysis together with the bioinformatics support at Institut Pasteur.</p>
</div>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p>Journal of Immunological Methods doi: <a href="https://doi.org/10.1016/j.jim.2021.113176" class="external-link">10.1016/j.jim.2021.113176</a></p>
<p><a href="https://www.biorxiv.org/content/10.1101/2020.06.07.127274v1" class="external-link uri">https://www.biorxiv.org/content/10.1101/2020.06.07.127274v1</a></p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>Developper tools need to be installed please see for further information:</p>
<p>Windows: <a href="https://cran.r-project.org/bin/windows/" class="external-link uri">https://cran.r-project.org/bin/windows/</a></p>
<p>MAC: <a href="https://cran.r-project.org/bin/macosx/" class="external-link uri">https://cran.r-project.org/bin/macosx/</a></p>
<div class="sourceCode" id="cb1"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">update.packages</span>()</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">"devtools"</span>))</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="st">"devtools"</span>)</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="co"># devtools::install_github("mul118/shinyMCE")</span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">"BiocManager"</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">"BiocManager"</span>)</span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="co"># update bioconductor packages if required</span></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>()</span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a><span class="co"># BiocManager::install("BiocSingular")</span></span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"InteractiveComplexHeatmap"</span>)</span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"SingleR"</span>) <span class="co"># needed to prepare data.</span></span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"haowulab/Wind"</span>)</span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"nghiavtr/BPSC"</span>)</span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"DEsingle"</span>)</span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"DESeq2"</span>)</span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"edgeR"</span>)</span>
<span id="cb1-17"><a href="#cb1-17" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"MAST"</span>)</span>
<span id="cb1-18"><a href="#cb1-18" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"monocle"</span>)</span>
<span id="cb1-19"><a href="#cb1-19" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"monocle3"</span>)</span>
<span id="cb1-20"><a href="#cb1-20" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"scDD"</span>)</span>
<span id="cb1-21"><a href="#cb1-21" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"limma"</span>)</span>
<span id="cb1-22"><a href="#cb1-22" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"Seurat"</span>)</span>
<span id="cb1-23"><a href="#cb1-23" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"statOmics/zingeR"</span>)</span>
<span id="cb1-24"><a href="#cb1-24" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"scater"</span>)</span>
<span id="cb1-25"><a href="#cb1-25" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"aggregation"</span>)</span>
<span id="cb1-26"><a href="#cb1-26" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"Zhangxf-ccnu/scDEA"</span>)</span>
<span id="cb1-27"><a href="#cb1-27" tabindex="-1"></a></span>
<span id="cb1-28"><a href="#cb1-28" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"InteractiveComplexHeatmap"</span>)</span>
<span id="cb1-29"><a href="#cb1-29" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"bookdown"</span>)</span>
<span id="cb1-30"><a href="#cb1-30" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span> (<span class="st">"GSVA"</span>)</span>
<span id="cb1-31"><a href="#cb1-31" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"GSEABase"</span>)</span>
<span id="cb1-32"><a href="#cb1-32" tabindex="-1"></a></span>
<span id="cb1-33"><a href="#cb1-33" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"BaderLab/Tempora"</span>)</span>
<span id="cb1-34"><a href="#cb1-34" tabindex="-1"></a></span>
<span id="cb1-35"><a href="#cb1-35" tabindex="-1"></a></span>
<span id="cb1-36"><a href="#cb1-36" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"C3BI-pasteur-fr/UTechSCB-SCHNAPPs"</span>, <span class="at">dependencies =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="create-sample-data-set">create sample data set<a class="anchor" aria-label="anchor" href="#create-sample-data-set"></a>
</h2>
<p>Load a small set of 200 PBMC cells and save to a file in the local directory. This file can be uploaded using the app.</p>
<p>A SingleCellExperiment object is needed with the following</p>
<p>colData: sampleNames <factor>, and barcode <factor></factor></factor></p>
<p>rowData: Description <character>, symbol<character>, id (unique)<character></character></character></character></p>
<div class="sourceCode" id="cb2"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"scEx"</span>, <span class="at">package =</span> <span class="st">"SCHNAPPs"</span>)</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="fu">save</span>(<span class="at">file =</span> <span class="st">"scEx.Rdata"</span>, <span class="at">list =</span> <span class="st">"scExOrg"</span>)</span></code></pre></div>
<div class="section level3">
<h3 id="history-functionality">history functionality<a class="anchor" aria-label="anchor" href="#history-functionality"></a>
</h3>
<p>To take advantage of the history functionality orca needs to be installed: (<a href="https://github.com/plotly/orca#installation" class="external-link uri">https://github.com/plotly/orca#installation</a>)</p>
<p>orca is part of ploty so nothing to be done for R. But the pdftools are required:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"pdftools"</span>)</span></code></pre></div>
<p>Notes:</p>
<ol style="list-style-type: decimal">
<li>once history check box is checked only the following plots will be recorded, the current plot will not be saved.</li>
<li>any selection in the plot will not be visible. It is not a screen-shot.</li>
<li>The date of recording is added to the title.</li>
<li>Creating of the history will take some time.</li>
<li>once activated, any changes of parameters will trigger re-plot and also a save to the history. This “feature” can also be used to create of the current plot by adding a “,” in one of the fields.</li>
</ol>
</div>
</div>
<div class="section level2">
<h2 id="running-schnapps">Running schnapps<a class="anchor" aria-label="anchor" href="#running-schnapps"></a>
</h2>
<p>To start the app:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">library</span>(SCHNAPPs)</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="fu">schnapps</span>()</span></code></pre></div>
<div class="section level3">
<h3 id="history-functionality-1">history functionality<a class="anchor" aria-label="anchor" href="#history-functionality-1"></a>
</h3>
<p>Plots can be automatically stored in a PDF file as they are created. This allows to somehow track what is being done. To enable this, the parameter historyFile has to be set to a file (probably non-existing, otherwise it will be appended to).</p>
</div>
<div class="section level3">
<h3 id="generate-data-files">generate data files<a class="anchor" aria-label="anchor" href="#generate-data-files"></a>
</h3>
<p>A singleCellExperiment object is required, saved in a file RData object using</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">save</span>(<span class="at">file =</span> <span class="st">"filename.RData"</span>, <span class="st">"singleCellExperiementObject"</span>)</span></code></pre></div>
<p>Please see <a href="https://c3bi-pasteur-fr.github.io/UTechSCB-SCHNAPPs/" class="external-link">GitHub</a> for further documentation on how to use schnapps.</p>
<p>Please see <a href="https://github.com/baj12/SCHNAPPsContributions" class="external-link">GitHub Contributions</a> for additional tools, not directly part of the SCHNAPPs pacakage.</p>
</div>
</div>
<div class="section level2">
<h2 id="extending-schapps">Extending SCHAPPs<a class="anchor" aria-label="anchor" href="#extending-schapps"></a>
</h2>
<p>See <a href="https://github.com/baj12/SCHNAPPsContributions" class="external-link">SCHNAPPsContribution</a> for examples and dummy contributions on how add functionality.</p>
</div>
<div class="section level2">
<h2 id="credits">Credits<a class="anchor" aria-label="anchor" href="#credits"></a>
</h2>
<p>The original version of this app is based on CellView (<a href="https://github.com/mohanbolisetty/CellView" class="external-link uri">https://github.com/mohanbolisetty/CellView</a>), but it was substantially modified. It helped me get started.</p>
<p>We are also greatful to all the members of the single cell working group at Pasteur Paris.</p>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/C3BI-pasteur-fr/UTechSCB-SCHNAPPs/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/C3BI-pasteur-fr/UTechSCB-SCHNAPPs/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>

<div class="community">
<h2 data-toc-skip>Community</h2>
<ul class="list-unstyled">
<li><a href="CONTRIBUTING.html">Contributing guide</a></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing SCHNAPPs</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Bernd Jagla <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-7696-0484" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Bernd Jagla.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
