<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Frequently asked questions for common tasks • SCHNAPPs</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><meta property="og:title" content="Frequently asked questions for common tasks">
<meta property="og:description" content="SCHNAPPs">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">SCHNAPPs</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.4.61</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/Readme.html">Readme</a>
    </li>
    <li>
      <a href="../../articles/pkdown/FAQs.html">Frequently asked questions for common tasks</a>
    </li>
    <li>
      <a href="../../articles/pkdown/SCHNAPPs_usage.html">SCHNAPPs - Single Cell sHiNy APP(s)</a>
    </li>
    <li>
      <a href="../../articles/pkdown/SeuratWorkflow.html">Seurat workflow</a>
    </li>
    <li>
      <a href="../../articles/pkdown/scranWorkflow.html">Follow scran workflow to analyze single-cell RNA-seq data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/C3BI-pasteur-fr/UTechSCB-SCHNAPPs/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="FAQs_files/header-attrs-2.2/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Frequently asked questions for common tasks</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/C3BI-pasteur-fr/UTechSCB-SCHNAPPs/blob/master/vignettes/pkdown/FAQs.Rmd"><code>vignettes/pkdown/FAQs.Rmd</code></a></small>
      <div class="hidden name"><code>FAQs.Rmd</code></div>

    </div>

    
    
<div id="scientific-questions" class="section level2">
<h2 class="hasAnchor">
<a href="#scientific-questions" class="anchor"></a>Scientific questions</h2>
<div id="identify-dead-cells" class="section level3">
<h3 class="hasAnchor">
<a href="#identify-dead-cells" class="anchor"></a>Identify Dead cells</h3>
<p>Dead cells can have high mitochondrial or ribosomal protein content. Using SCHNAPPs those can be identified before and after removing of such genes from the data set. Though they have to be present in the original input data set.</p>
<div id="before-removing-genes" class="section level4">
<h4 class="hasAnchor">
<a href="#before-removing-genes" class="anchor"></a>before removing genes</h4>
<div class="figure" style="text-align: center">
<img src="images/beforeFilter.png" alt="Set before-Filter variable" width="90%"><p class="caption">
Set before-Filter variable
</p>
</div>
<p>The regular expression in the image above is used to count UMIs in the original non-normalized data before any filters are applied. Then we have this count for the cells that have not been removed and can use this in a 2D plot:</p>
<div class="figure" style="text-align: center">
<img src="images/beforeFilterApplied.png" alt="Before filter applied" width="90%"><p class="caption">
Before filter applied
</p>
</div>
<p>Here, we show the UMI count on the x-axis and the before filter on the y-axis. We also devide the UMI count by the before filter column, which emphasises the effect and allows us to clearly identify dying cells. In this case the dying cells fall into clusters 1, 2, and 3. So we can select those clusters and copy/paste the cells in the “cells to be removed field”.</p>
<div class="figure" style="text-align: center">
<img src="images/beforeFilterCellSelect.png" alt="Select cells for copy/paste" width="90%"><p class="caption">
Select cells for copy/paste
</p>
</div>
<div class="figure" style="text-align: center">
<img src="images/beforeFilterCellsRM.png" alt="remove cells" width="90%"><p class="caption">
remove cells
</p>
</div>
<p>Don’t forget to click on apply changes!</p>
</div>
</div>
<div id="identify-cells-from-violin-plot" class="section level3">
<h3 class="hasAnchor">
<a href="#identify-cells-from-violin-plot" class="anchor"></a>Identify cells from violin plot</h3>
<p>The cells in the violin plot are not selectable (for the moment). In order to identify those cells one has to use the 2D plotting function, preferably under Co-expression - Selected</p>
</div>
<div id="can-mars-seq-data-be-analyzed" class="section level3">
<h3 class="hasAnchor">
<a href="#can-mars-seq-data-be-analyzed" class="anchor"></a>Can MARS-Seq data be analyzed?</h3>
<p>Yes, additional annotations can be either directly into a singlecellexperiment object that is loaded in the app as colData, or using a CSV file with corresponding information.</p>
<p>Use</p>
<blockquote>
<p>colData(scEx) &lt;- myAdditionalInformation</p>
</blockquote>
<p>to set the relevant information.</p>
</div>
<div id="how-to-normalize-by-ribosomal-proteins" class="section level3">
<h3 class="hasAnchor">
<a href="#how-to-normalize-by-ribosomal-proteins" class="anchor"></a>How to normalize by ribosomal proteins?</h3>
<p>Once you have identified the gene names you want to use this list of genes can be used under Parameters - Normalization - gene_norm:</p>
<div class="figure" style="text-align: center">
<img src="images/normByRP.png" alt="normalize by ribosomal proteins" width="90%"><p class="caption">
normalize by ribosomal proteins
</p>
</div>
<p>You can get this list by temporarily setting the regular expression for selection of genes to be removed. Then click apply changes.</p>
<div class="figure" style="text-align: center">
<img src="images/geneSelectionRegEx.png" alt="regular expression for selecting ribosomal/mitochondrial genes" width="90%"><p class="caption">
regular expression for selecting ribosomal/mitochondrial genes
</p>
</div>
<p>Then, in the lower area of the same screen the genes that have been removed are shown:</p>
<div class="figure" style="text-align: center">
<img src="images/geneSelectionRMGenes.png" alt="table of removed genes" width="90%"><p class="caption">
table of removed genes
</p>
</div>
</div>
<div id="what-is-the-best-representation-to-use-to-compare-gene-expression-level-umigene" class="section level3">
<h3 class="hasAnchor">
<a href="#what-is-the-best-representation-to-use-to-compare-gene-expression-level-umigene" class="anchor"></a>What is the best representation to use to compare gene expression level ( UMI/gene) ?</h3>
<p>Gene/UMI expression can be visualized in different plots:</p>
<ul>
<li>The <b>panel plot</b> gives an overview of which cells express different genes projected in 2 dimensions. The dimensions can be chosen freely. This allows for a general overview of the regions in the PCA/tSNE/UMAP space where differences are expressed. It can be difficult to compare the expression of individual cells.</li>
</ul>
<div class="figure" style="text-align: center">
<img src="images/DataexplorationPanel.png" alt="Panel plot" width="90%"><p class="caption">
Panel plot
</p>
</div>
<ul>
<li>Using the parameters UmiCountPerGenes and UmiCountPerGenes2 allow plotting the expression of one or more genes (i.e. the sum of all UMIs) that are set with these variables. These two parameters can be found under X/Y when selecting the projections to be used for the X/Y axes. With these parameters, the expression of two sets of genes can be compared and the information of individual cells can be retained. Zooming and coloring allows drilling down and identifying individual cells. Using “more options” these can then be annotated. Following such annotation, this information can be used in other plots as a projection. Those projections are currently binary, i.e. have the values False and True or 0 and 1, indicating if a cell belongs to a group or not.</li>
</ul>
<p>The figure below shows the combined expression of marker genes for B and T cells in an PBMC experiment. Note that the first gene name is B-cells:, This will allow you to annotate the selection. It will give rise to a warning, but that will disappear after a few seconds.</p>
<div class="figure" style="text-align: center">
<img src="images/cd8vscd4.png" alt="2D plot CD8 vs CD4 cell expression" width="90%"><p class="caption">
2D plot CD8 vs CD4 cell expression
</p>
</div>
</div>
<div id="how-to-evaluate-if-the-clusters-are-biologically-relevant" class="section level3">
<h3 class="hasAnchor">
<a href="#how-to-evaluate-if-the-clusters-are-biologically-relevant" class="anchor"></a>How to evaluate if the clusters are biologically relevant?</h3>
<iframe src="https://www.youtube.com/embed/kL00ceA51vs" width="420" height="315" frameborder="0" allowfullscreen=""></iframe>
</div>
<div id="how-to-identify-cells-in-a-cluster-of-a-tsne-plot-and-list-them" class="section level3">
<h3 class="hasAnchor">
<a href="#how-to-identify-cells-in-a-cluster-of-a-tsne-plot-and-list-them" class="anchor"></a>How to identify cells in a cluster of a Tsne plot and list them ?</h3>
<p>Using the 2D plot, select tsne1 and tsne2. Using the mouse select the cells of interest. Check “more options”, then check “show cell names”.</p>
<div class="figure" style="text-align: center">
<img src="images/CellSelection2.png" alt="selecting cells" width="90%"><p class="caption">
selecting cells
</p>
</div>
<!----
### What are the different heat maps that can be generated and how many genes could be listed  by cluster ?

Cells in the cluster can be selected on the 2D plot and cell names are shown after checking “show more options” followed by checking “show cell names”; 

---->
</div>
<div id="what-is-the-best-representation-to-use-to-compare-gene-expression-level" class="section level3">
<h3 class="hasAnchor">
<a href="#what-is-the-best-representation-to-use-to-compare-gene-expression-level" class="anchor"></a>What is the best representation to use to compare gene expression level</h3>
<p>Gene/UMI expression can be visualized in different plots:</p>
<p>The <b>panel plot</b> gives an overview which cells express different genes projected in 2 dimensions. The dimensions can be chosen freely. This allows for a general overview of the regions in the PCA/tSNE/UMAP space where differences are expressed. It can be difficult to compare the expression of individual cells.</p>
<p>Using <b>2D plots</b>, the parameters UmiCountPerGenes and UmiCountPerGenes2 allow plotting the expression of one or more genes (i.e. the sum of all UMIs) that are set with these variables. These two parameters can be found under X/Y when selecting the projections to be used for the X/Y axes. With these parameters, the expression of two sets of genes can be compared and the information of individual cells can be retained. Zooming and coloring allows drilling down and identifying individual cells. Using “more options” these can then be annotated. Following such annotation, this information can be used in other plots as a projection. Those projections are currently binary, i.e. have the values False and True or 0 and 1, indicating if a cell belongs to a group or not.</p>
</div>
<div id="can-the-analysis-be-performed-on-a-subset-of-selected-cells-and-how" class="section level3">
<h3 class="hasAnchor">
<a href="#can-the-analysis-be-performed-on-a-subset-of-selected-cells-and-how" class="anchor"></a>Can the analysis be performed on a subset of selected cells and how?</h3>
<p>Analysis can be narrowed to a selection of cells copying cell names in the tab “Cell selection - Cells to keep, remove others”. The number of cells will be updated in the summary statistics on the sidebar and clicking on “apply changes” will apply cell selection.</p>
<!----
### How to identify signature genes with same distribution as a given marker gene? 

--->
</div>
<div id="what-are-the-possible-trajectory-inferences-approaches-that-could-be-used" class="section level3">
<h3 class="hasAnchor">
<a href="#what-are-the-possible-trajectory-inferences-approaches-that-could-be-used" class="anchor"></a>What are the possible trajectory inferences approaches that could be used ?</h3>
<p>Contributions for trajectory inference already available are SCORPIUS, ElPiGraph.R, and DCA (see GitHub)[<a href="https://github.com/baj12/SCHNAPPsContributions" class="uri">https://github.com/baj12/SCHNAPPsContributions</a>]</p>
<!----

### To display a developmental pathway according to time of expression of genes from an input involving several datasets ?


### To display a biological  process from a single dataset ?
--->
</div>
<div id="can-new-packages-be-implemented" class="section level3">
<h3 class="hasAnchor">
<a href="#can-new-packages-be-implemented" class="anchor"></a>Can new packages be implemented ?</h3>
<p>The software architecture of the application makes it easy for the bioinformatician to integrate new visualizations and analyses. please see <a href="https://github.com/baj12/SCHNAPPsContributions" class="uri">https://github.com/baj12/SCHNAPPsContributions</a></p>
</div>
</div>
<div id="common-problems" class="section level2">
<h2 class="hasAnchor">
<a href="#common-problems" class="anchor"></a>Common problems</h2>
<div id="nothing-happens" class="section level3">
<h3 class="hasAnchor">
<a href="#nothing-happens" class="anchor"></a>Nothing happens!!!!!</h3>
<p>Patience, please. There are quite a few occasions when the user has to wait. Specifically, when opening the app on a fresh start. There are some 100 packages or so to be loaded. This takes time.</p>
<p>Other places that take long waiting times are after loading data (then most calculations will be initialized), changing the parameters for normalization, PCA, … (everything that touches the normalized data and projections).</p>
<p>UMAP is notoriously slow, that is why there is to check box to activate it.</p>
<p>Please see if the processors on your computer related to R are active, this usually means that you should wait.</p>
<p>There is no clean way stopping a running task other than killing it.</p>
<p>Reports take even longer. Wait times of an hour or more are not uncommon. Then after a screen for saving the file appears and Ok is pressed the same time will be used again to redo the report. I haven’t found a way yet to overcome this. Any ideas are welcome.</p>
</div>
<div id="new-windows-pop-up" class="section level3">
<h3 class="hasAnchor">
<a href="#new-windows-pop-up" class="anchor"></a>New windows pop up</h3>
<p>When saving files or generating a report Shiny opens a new window and doesn’t close it. You can safely close it after the operation is completed.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Bernd Jagla.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
