<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>SCHNAPPs - Single Cell sHiNy APP(s) • SCHNAPPs</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><meta property="og:title" content="SCHNAPPs - Single Cell sHiNy APP(s)">
<meta property="og:description" content="SCHNAPPs">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">SCHNAPPs</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.4.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/Readme.html">Readme</a>
    </li>
    <li>
      <a href="../../articles/pkdown/FAQs.html">Frequently asked questions for common tasks</a>
    </li>
    <li>
      <a href="../../articles/pkdown/SCHNAPPs_usage.html">SCHNAPPs - Single Cell sHiNy APP(s)</a>
    </li>
    <li>
      <a href="../../articles/pkdown/children/child.tartour.anova.html">UNKNOWN TITLE</a>
    </li>
    <li>
      <a href="../../articles/pkdown/customReportFromHistory.html">Custom Report from History</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/C3BI-pasteur-fr/UTechSCB-SCHNAPPs/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>SCHNAPPs - Single Cell sHiNy APP(s)</h1>
            
            <h4 class="date">2020-08-25</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/C3BI-pasteur-fr/UTechSCB-SCHNAPPs/blob/master/vignettes/pkdown/SCHNAPPs_usage.Rmd"><code>vignettes/pkdown/SCHNAPPs_usage.Rmd</code></a></small>
      <div class="hidden name"><code>SCHNAPPs_usage.Rmd</code></div>

    </div>

    
    
<div id="overview" class="section level2">
<h2 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h2>
<p>Shiny app for the exploration and analysis of single cell RNAseq data as it comes from 10X or MARSseq technologies or other. It is currently being developed based on user requests of the Cytometry and Biomarkers UTechS at the Institut Pasteur, Paris. The goal is to enable the users of our platform to explore their data, select cells they would like to work with and then perform the final analysis together with the bioinformatics support at Pasteur. We hope you might find it helpful as well.</p>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<pre><code>if (!require("devtools"))
  install.packages("devtools")
devtools::install_github("mul118/shinyMCE")
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("BiocSingular")
install_github("C3BI-pasteur-fr/UTechSCB-SCHNAPPs")</code></pre>
</div>
<div id="running-schnapps" class="section level2">
<h2 class="hasAnchor">
<a href="#running-schnapps" class="anchor"></a>Running schnapps</h2>
<p>The application is started from the command line in R using schnapps().</p>
<pre><code><a href="https://rdrr.io/pkg/SCHNAPPs/man">library(SCHNAPPs)
# create example data
save(file = "filename.RData", "singleCellExperiementObject")
# use "filename.RData" with load data functionality within the shiny app
schnapps()</a></code></pre>
<p>Optional parameters:</p>
<ul>
<li>
<p>localContributionDir</p>
<p>Additional functionally can be added using a vector with the paths to the contributions directories. See for example: (<a href="https://github.com/baj12/scShinyHubContributionsBJ" class="uri">https://github.com/baj12/scShinyHubContributionsBJ</a>)</p>
</li>
<li>
<p>defaultValueSingleGene</p>
<p>Some of the UI elements expect a single gene as input. This parameter can be preset with this value.</p>
</li>
<li>
<p>defaultValueMultiGenes</p>
<p>Some of the UI elements expect multiple genes as input. This parameter can be preset with this value.</p>
</li>
<li>
<p>defaultValueRegExGene</p>
<p>Default parameters for the regular expression used in the Gene selection tab (defaults to <sup>MT-|</sup>RP|^MRP</p>
</li>
<li>
<p>DEBUG</p>
<p>If set to TRUE, debugging information will be displayed on the console</p>
</li>
<li>
<p>DEBUGSAVE</p>
<p>If set to TRUE, intermediate file will be written to disc. This delays calculation. It also creates a directory in the home folder called SCHNAPPsDEBUG, where all saved files are stored.</p>
</li>
</ul>
<div id="loadData" class="section level3">
<h3 class="hasAnchor">
<a href="#loadData" class="anchor"></a>Load data</h3>
<p>To load count data there are two formats that are accepted:</p>
<ul>
<li>singleCellExperiment object (preferred)</li>
<li>CSV file with genes per row and cells as columns.</li>
</ul>
<p>In addition cell or gene metadata can be loaded from a CSV file (optional)</p>
<p>A singleCellExperiment object is required, saved in a file RData object using</p>
<pre><code><a href="https://rdrr.io/r/base/save.html">save(file = "filename.RData", "singleCellExperiementObject")</a></code></pre>
<p>Load a small set of 200 cells and save to a file in the local directory</p>
<pre><code><a href="https://rdrr.io/r/utils/data.html">data("scEx", package = "SCHNAPPs")
save(file = "scEx.Rdata", list = "scEx")</a></code></pre>
</div>
</div>
<div id="overview-of-functionality" class="section level2">
<h2 class="hasAnchor">
<a href="#overview-of-functionality" class="anchor"></a>Overview of functionality</h2>
<p>Below is the description of the individual tabs/views with the SCHNAPPs application. During the execution of a task, a message is displayed at the lower right corner. While the message is displayed other requests will wait.</p>
<div class="figure" style="text-align: center">
<img src="images/overview.png" alt="Overview, showing entry screen" width="90%"><p class="caption">
Overview, showing entry screen
</p>
</div>
<div id="side-panel" class="section level3">
<h3 class="hasAnchor">
<a href="#side-panel" class="anchor"></a>Side panel</h3>
<ul>
<li>
<p>Tabs: input, Parameters, Cell selection, Gene selection</p>
<p>See below.</p>
</li>
<li>
<p>Summary statistics of this dataset</p>
<p>The name of the input files are listed, number of cells that are actively used (not necessarily the input data, but after applying the filters), number of genes with more than one UMI over all cells, median UMIs per cell, median number of genes with more than 1 UMI per cell, total number of UMIs (reads), memory used (just a very rough estimate at the beginning of the calculations), normalization method selected in parameters.</p>
</li>
<li>
<p>Generate report</p>
<p>An HTML report is created when clicking on this button. All visualizations are stored in that report with all parameters. Several other files are also stored. Everything is compiled in a ZIP file. Included are the list of differentially expressed genes, the projections as CSV, normalized counts as CSV, the table of Co-expressing genes (from Co-expression - selected), a readme.txt file, the report.html and the sessionData.RData, which includes the data for reanalysis in R as well as it can be again read into the application. Be aware that creating this report might take some time.</p>
</li>
<li>
<p>Download counts.csv</p>
<p>Download the normalized data as a CSV file.</p>
</li>
<li>
<p>Download Rdata</p>
<p>Download an RData file with the scEx object that can be reloaded in SCHNAPPs or other applications (e.g. iSEE).</p>
</li>
<li>
<p>(Save for DEBUG)</p>
<p>Only displayed when DEBUG is set to TRUE during start-up. Can be selected to save internal data for debugging purposes. This delays calculation. It also creates a directory in the home folder called SCHNAPPsDEBUG, where all saved files are stored.</p>
</li>
</ul>
<div class="figure" style="text-align: center">
<img src="images/loadingData.png" alt="While loading data" width="90%"><p class="caption">
While loading data
</p>
</div>
</div>
<div id="tabs" class="section level3">
<h3 class="hasAnchor">
<a href="#tabs" class="anchor"></a>Tabs</h3>
<div id="input" class="section level4">
<h4 class="hasAnchor">
<a href="#input" class="anchor"></a>Input</h4>
<ul>
<li>
<p>Upload annotation for cells and/or genes</p>
<p>Examples for such files are given in :scExGenes.csv (gene annotations), scExCells.csv (cell annotations). Also the output from scorpius (see contributions) could be used here.</p>
</li>
<li>
<p>Upload RData or csv file</p>
<p>The RData file has to contain a singleCellExperiment object called scEx. Only the count matrix and the rowData/colData are used as the transformations and projections will be recalculated based on the actual cells used. An example for the RData file is scEx.RData. To create the data locally follow the instructs under <a href="#loadData">“Load data”</a>.</p>
</li>
<li>
<p>regular expression to count genes/cell</p>
<p>This is a regular expression that is calculate a projection called “before.filter”. These counts are based on the original input data. A common use case would be to include here ribosomal proteins and use this in a 2D plot together with UMI counts to visualize dying cells.</p>
</li>
</ul>
</div>
<div id="parameters" class="section level4">
<h4 class="hasAnchor">
<a href="#parameters" class="anchor"></a>Parameters</h4>
<div id="normalization" class="section level5">
<h5 class="hasAnchor">
<a href="#normalization" class="anchor"></a>Normalization</h5>
<div class="figure" style="text-align: center">
<img src="images/parametersNormalization.png" alt="Parameters - Normalization" width="90%"><p class="caption">
Parameters - Normalization
</p>
</div>
<ul>
<li>
<p>rawNormalization:</p>
<p>Umi counts, no transformation performed. Counts will still end up in the log-transformed slot</p>
</li>
<li>
<p>scEx-Log</p>
<p>Normalize by the total number of UMIs, log2(x+1) transform, scale with factor 1000</p>
</li>
<li>
<p>gene_norm</p>
<p>Divide each expression by the sum of the genes supplied per cell. If no genes supplied then it is the same as scEx-log</p>
<p>Additional parameters for a normalization from a contribution will be shown here as well. Since the standard normalizations don’t have any, none is shown.</p>
</li>
<li>
<p>Cells table</p>
<p>A table with the normalized values is shown below. Only 20 cells are shown (columns). Table shows the filtered cells/genes (after applying the filters from Cell selection and Gene selection). The rows can be sorted by clicking on the cell name. Individual genes can be selected by clicking in the row. Those genes are shown above the table (for copy/paste actions). The cells (all not only the 20 shown) can be sorted by the sum of the normalized gene expression when the corresponding check box is selected. The cellnames cannot be selected/copied. The full table can also be downloaded as a CSV (comma seperated values) file.</p>
</li>
</ul>
</div>
<div id="general-parameters" class="section level5">
<h5 class="hasAnchor">
<a href="#general-parameters" class="anchor"></a>General Parameters</h5>
<div class="figure" style="text-align: center">
<img src="images/parametersGeneral.png" alt="Parameters - General" width="90%"><p class="caption">
Parameters - General
</p>
</div>
<p>Quickcluster from the scran package is used for clustering of cells.</p>
<ul>
<li>
<p>Parameters for clustering</p>
<ul>
<li>
<p>use PCA or normalized data?</p>
<ul>
<li>
<p>PCA (default)</p>
<p>Use the PCA components for the normalization</p>
</li>
<li>
<p>normData</p>
<p>Use all data for normalization. This can take substantially more time and memory.</p>
</li>
</ul>
</li>
<li>
<p>minimum size of each cluster.</p>
<p>The minimum size of each cluster.</p>
</li>
<li>
<p>clustering method to use</p>
<ul>
<li>
<p>Igraph (default)</p>
<p>A shared nearest neighbor graph is constructed using the buildSNNGraph function. This is used to define clusters based on highly connected communities in the graph, using the graph.fun function.</p>
</li>
<li>
<p>Hclust</p>
<p>a distance matrix is constructed; hierarchical clustering is performed using Ward’s criterion; and cutreeDynamic is used to define clusters of cells. This can take substantially more time and memory.</p>
</li>
</ul>
</li>
<li>
<p>Genes to be used for clustering</p>
<p>Only genes shown here will be used for normalization. The normalization factor will be applied to all genes. If left empty all genes (after filtering) will be used.</p>
</li>
</ul>
</li>
<li>
<p>Comments</p>
<p>The work can be commented. These comments will show with the formatting in the report when “Generate report” is clicked.</p>
</li>
<li><p>Colors The colors for samples and cluster can be modified. The “Update colours” button needs to be clicked to apply the canges.</p></li>
</ul>
</div>
<div id="tsne-plot" class="section level5">
<h5 class="hasAnchor">
<a href="#tsne-plot" class="anchor"></a>TSNE plot</h5>
<div class="figure" style="text-align: center">
<img src="images/parametersTSNE.png" alt="Parameters - tSNE" width="90%"><p class="caption">
Parameters - tSNE
</p>
</div>
<p>Place to change the parameters for the tSNE calculations (Wrapper for the C++ implementation of Barnes-Hut t-Distributed Stochastic Neighbor Embedding. t-SNE is a method for constructing a low dimensional embedding of high-dimensional data, distances or similarities. Exact t-SNE can be computed by setting theta=0.0.) . Rtsne package is used. All PCA components available are used for the calculations. The number of PCs calculated is set under Parameters - General Parameters.</p>
<p>The 3D display is not limited to the tsne projections and thus can be used to</p>
<ul>
<li>
<p>Tsne dimensions</p>
<p>The number of dimensions to be calculated for tSNE. Fixe to 3 currently.</p>
</li>
<li>
<p>Perplexity</p>
<p>Perplexity parameter (should not be bigger than 3 * perplexity &lt; nrow(X) - 1, see details Rtsne for further information) default = 30, min = 1, max = 100.</p>
</li>
<li>
<p>Theta</p>
<p>Speed/accuracy trade-off (increase for less accuracy), set to 0.0 for exact TSNE (default: 0.5), min = 0.0, max = 1, step = 0.1.</p>
</li>
<li>
<p>Seed</p>
<p>Seed for random data generator to be used. Default =1, min = 1, max = 10000.</p>
</li>
<li>
<p>X</p>
Dimension for the X axis.</li>
<li>
<p>Y</p>
<p>Dimension for the Y axis.</p>
</li>
<li>
<p>Z</p>
<p>Dimension for the Z axis.</p>
</li>
<li>
<p>Colored by</p>
<p>Dimension for the color to be used.</p>
</li>
<li>
<p>Table</p>
<p>Table of all projections</p>
</li>
</ul>
</div>
<div id="umap" class="section level5">
<h5 class="hasAnchor">
<a href="#umap" class="anchor"></a>Umap</h5>
<div class="figure" style="text-align: center">
<img src="images/parametersUMAP.png" alt="Parameters - UMAP" width="90%"><p class="caption">
Parameters - UMAP
</p>
</div>
<p>Calculation of Umap projections. Since this can be quite time consuming there is a checkbox called “activate Umap projection” that starts the computation. When this is selected all changes in the parameters result in new executions. Thus it should be unchecked while the parameters are changed. The R package uwot is used. The description of the parameters comes directly from the package documentation.</p>
<ul>
<li>
<p>random seed</p>
<p>Seed for random data generator to be used. Default =1, min = 1, max = 100.</p>
</li>
<li>
<p>N Neighbors</p>
<p>The size of local neighborhood (in terms of number of neighboring sample points) used for manifold approximation. Larger values result in more global views of the manifold, while smaller values result in more local data being preserved. In general values should be in the range 2 to 100.</p>
</li>
<li>
<p>N components</p>
<p>The dimension of the space to embed into. This defaults to 2 to provide easy visualization, but can reasonably be set to any integer value in the range 2 to 100.</p>
</li>
<li>
<p>negative sample rate</p>
<p>The number of negative edge/1-simplex samples to use per positive edge/1-simplex sample in optimizing the low dimensional embedding.</p>
</li>
<li>
<p>Metric</p>
<p>Type of distance metric to use to find nearest neighbors. Possible values: “euclidean”, “manhattan”, “cosine”, “hamming”</p>
</li>
<li>
<p>Epochs</p>
<p>Number of epochs to use during the optimization of the embedded coordinates. By default, this value is set to 500 for datasets containing 10,000 vertices or less, and 200 otherwise.</p>
</li>
<li>
<p>Init</p>
<p>Type of initialization for the coordinates. Possible values: “spectral”, “random”.</p>
</li>
<li>
<p>Spread</p>
<p>The effective scale of embedded points. In combination with min_dist, this determines how clustered/clumped the embedded points are.</p>
</li>
<li>
<p>min dist</p>
<p>The effective minimum distance between embedded points. Smaller values will result in a more clustered/clumped embedding where nearby points on the manifold are drawn closer together, while larger values will result on a more even dispersal of points. The value should be set relative to the spread value, which determines the scale at which embedded points will be spread out.</p>
</li>
<li>
<p>Set op mix ratio</p>
<p>Interpolate between (fuzzy) union and intersection as the set operation used to combine local fuzzy simplicial sets to obtain a global fuzzy simplicial sets. Both fuzzy set operations use the product t-norm. The value of this parameter should be between 0.0 and 1.0; a value of 1.0 will use a pure fuzzy union, while 0.0 will use a pure fuzzy intersection.</p>
</li>
<li>
<p>local connectivity</p>
<p>The local connectivity required – i.e. the number of nearest neighbors that should be assumed to be connected at a local level. The higher this value the more connected the manifold becomes locally. In practice this should be not more than the local intrinsic dimension of the manifold.</p>
</li>
<li>
<p>Bandwidth</p>
<p>The effective bandwidth of the kernel if we view the algorithm as similar to Laplacian Eigenmaps. Larger values induce more connectivity and a more global view of the data, smaller values concentrate more locally.</p>
</li>
</ul>
</div>
</div>
<div id="general-qc" class="section level4">
<h4 class="hasAnchor">
<a href="#general-qc" class="anchor"></a>General QC</h4>
<div id="umi-histogram" class="section level5">
<h5 class="hasAnchor">
<a href="#umi-histogram" class="anchor"></a>UMI histogram</h5>
<div class="figure" style="text-align: center">
<img src="images/generalUMI.png" alt="General QC - UMI histogram" width="90%"><p class="caption">
General QC - UMI histogram
</p>
</div>
<p>The number of UMIs per cell are displayed as a histogram. Different samples are colored differently. This allows identifying thresholds to be used in the cell selection panel.</p>
</div>
<div id="sample-histogram" class="section level5">
<h5 class="hasAnchor">
<a href="#sample-histogram" class="anchor"></a>Sample histogram</h5>
<div class="figure" style="text-align: center">
<img src="images/generalSample.png" alt="General QC - sample histogram" width="90%"><p class="caption">
General QC - sample histogram
</p>
</div>
<p>Histgram of cells per samples. This allows verifying that the number of cells per sample is comparable.</p>
</div>
<div id="pc-variance" class="section level5">
<h5 class="hasAnchor">
<a href="#pc-variance" class="anchor"></a>PC variance</h5>
<div class="figure" style="text-align: center">
<img src="images/generalPC.png" alt="General QC - PC variances" width="90%"><p class="caption">
General QC - PC variances
</p>
</div>
<p>Variance of the first 10 principle components of the PCA are shown. This allows verifying the importance of the different PCs.</p>
</div>
<div id="scater-qc" class="section level5">
<h5 class="hasAnchor">
<a href="#scater-qc" class="anchor"></a>Scater QC</h5>
<div class="figure" style="text-align: center">
<img src="images/generalScater1.png" alt="General QC - highest expressed genes before checking run" width="90%"><p class="caption">
General QC - highest expressed genes before checking run
</p>
</div>
<div class="figure" style="text-align: center">
<img src="images/generalScater2.png" alt="General QC - highest expressed genes" width="90%"><p class="caption">
General QC - highest expressed genes
</p>
</div>
<p>Plots the highest expressing genes based on the scater package. (scater::plotHighestExprs), colour_cells_by = “log10_total_counts” and a maximum of 50 genes is displayed. This can take quite some time to compute for larger (or even small) data sets.</p>
</div>
</div>
<div id="cell-selection" class="section level4">
<h4 class="hasAnchor">
<a href="#cell-selection" class="anchor"></a>Cell selection</h4>
<div class="figure" style="text-align: center">
<img src="images/CellSelection.png" alt="Cell selection" width="90%"><p class="caption">
Cell selection
</p>
</div>
<p>All selections are based on the original input data without filtering genes. In case this is needed, one can apply the gene filter, save the RData file and then apply the cell filters.</p>
<ul>
<li>
<p>List of genes with minimal expression</p>
<p>Specify a comma separated list of genes that have to be expressed with at least one UMI. If you want to restrict to cells that express a gene to more than one UMI, you would have to select those cells manually using one of the 2D plots.</p>
</li>
<li>
<p>Min # of UMIs / Max # of UMIs</p>
<p>Select only cells that have at least / no more than X number of UMIs.</p>
</li>
<li>
<p>Comment for selection of cells</p>
<p>The text entered here will be displayed in the report</p>
</li>
<li><p>cells to be filtered out by pattern</p></li>
<li><p>cells to keep</p></li>
<li><p>cells to keep; remove others</p></li>
<li><p>Cells to be removed</p></li>
<li>
<p>Cell table</p>
<p>All cells are shown, even those that have been removed. The result of the removal can be verified in the stats section of the side-panel.</p>
</li>
</ul>
</div>
<div id="gene-selection" class="section level4">
<h4 class="hasAnchor">
<a href="#gene-selection" class="anchor"></a>Gene selection</h4>
<img src="images/geneSelection1.png" width="90%" style="display: block; margin: auto;"><div class="figure" style="text-align: center">
<img src="images/geneSelection2.png" alt="General QC - highest expressed genes" width="90%"><p class="caption">
General QC - highest expressed genes
</p>
</div>
<ul>
<li><p>regular expression for selection of genes to be removed</p></li>
<li>
<p>GeneList Selection</p>
<p>From this list sets of genes belonging to a specific class of proteins can be selected. This is only useful for human genes. Specific tables for other genomes could be made available upon request.</p>
</li>
<li>
<p>Min expression over all cells</p>
<p>This corresponds to the rowSums in the kept genes table and refers total number of UMIs for a given gene over all cells.</p>
</li>
<li>
<p>genes to keep</p>
<p>Set the gene names that should be kept.</p>
</li>
<li>
<p>Table Genes kept</p>
<p>Holds the genes that are left after removal. The column rowSums holds the sum of UMIs for all cells The column rowSamples holds the number of cells where the gene is expressed</p>
</li>
<li>
<p>Table Genes removed</p>
<p>Shows the genes that were removed.</p>
</li>
</ul>
</div>
<div id="co-expression" class="section level4">
<h4 class="hasAnchor">
<a href="#co-expression" class="anchor"></a>Co-expression</h4>
<div id="all-clusters" class="section level5">
<h5 class="hasAnchor">
<a href="#all-clusters" class="anchor"></a>All clusters</h5>
<div class="figure" style="text-align: center">
<img src="images/coexpressionAll.png" alt="Co-expression - All clusters" width="90%"><p class="caption">
Co-expression - All clusters
</p>
</div>
<ul>
<li>
<p>Comma seperated gene names</p>
<p>Comma separated list of genes to be displayed in the heatmap for all cells.</p>
</li>
<li>
<p>Heatmap plot</p>
<p>Uses the heatmap module see Modules Heatmap in this document</p>
</li>
</ul>
</div>
<div id="selected" class="section level5">
<h5 class="hasAnchor">
<a href="#selected" class="anchor"></a>Selected</h5>
<div class="figure" style="text-align: center">
<img src="images/coexpressionSelected1.png" alt="Co-expression - Selected -- before selection" width="90%"><p class="caption">
Co-expression - Selected – before selection
</p>
</div>
<img src="images/coexpressionSelected2.png" width="90%" style="display: block; margin: auto;"><div class="figure" style="text-align: center">
<img src="images/coexpressionSelected3.png" alt="Co-expression - Selected" width="90%"><p class="caption">
Co-expression - Selected
</p>
</div>
<ul>
<li>
<p>2D plot</p>
<p>Uses a 2D plot module to display cells. Selected cells are being used in heatmap and table below.</p>
</li>
<li>
<p>Heatmap</p>
<p>Uses the selected cells from the above 2D Plot and displays a heat map.</p>
</li>
<li>
<p>min UMI count per gene:</p>
<p>For the table below, set the minimum number of UMIs for a given gene to be considered.</p>
</li>
<li>
<p>min percentage of cells expressing a genes</p>
<p>Genes in the table below have to be expressed in at least this percent of the cells</p>
</li>
<li>
<p>Table of “co-expressing” genes</p>
<p>Uses the Table module to display for the selected cells the genes that are expressed with a minimal UMI count (see above) in at least X % (see above) of the cells. This is yet another way to identify genes that are expressed over all the selected cells.</p>
</li>
</ul>
</div>
<div id="violin-plot" class="section level5">
<h5 class="hasAnchor">
<a href="#violin-plot" class="anchor"></a>Violin plot</h5>
<div class="figure" style="text-align: center">
<img src="images/coexpressionViolin1.png" alt="Co-expression - Violin plot" width="90%"><p class="caption">
Co-expression - Violin plot
</p>
</div>
<div class="figure" style="text-align: center">
<img src="images/coexpressionViolin2.png" alt="Co-expression - Violin plot with permutations" width="90%"><p class="caption">
Co-expression - Violin plot with permutations
</p>
</div>
<ul>
<li>
<p>show Permutations</p>
<p>Selecting this can cause substantial calculations if the number of genes is large and make the display very ugly. On the othe hand, it will show which combination of genes are co-expressed to which extend.</p>
</li>
<li>
<p>Comma seperated gene names</p>
<p>A comma separated list of genes to investigate using the violin plot. If a gene doesn’t show it is not expressed in any of the cells.</p>
</li>
<li>
<p>X</p>
<p>X-axis to be used. Only factors (cell attributes with less than 21 unique values) are shown.</p>
</li>
<li>
<p>min expression of genes</p>
<p>Only genes with at least this number of UMIs are considered for the plot. This allows one to remove potential artefacts. Normalized counts are used.</p>
</li>
<li>
<p>Violin plot</p>
<p>Cannot be interogated. This gives just an overview. The cells coresponding to the individual values have to be selected using one of the 2D plot (preferably the Co-expression-selected)</p>
</li>
</ul>
</div>
<div id="som-cluster" class="section level5">
<h5 class="hasAnchor">
<a href="#som-cluster" class="anchor"></a>SOM cluster</h5>
<div class="figure" style="text-align: center">
<img src="images/coexpressionSOM1.png" alt="Co-expression - SOM network too large" width="90%"><p class="caption">
Co-expression - SOM network too large
</p>
</div>
<div class="figure" style="text-align: center">
<img src="images/coexpressionSOM2.png" alt="Co-expression - SOM" width="90%"><p class="caption">
Co-expression - SOM
</p>
</div>
<p>A self organizing map is calculated of all the cells. The matrix of normalized counts is used for the clustering. A square SOM with X nodes per dimension is calculated using nEpoch = 10, radius0 = 0, radiusN = 0, radiusCooling = “linear”, mapType = “planar”, gridType = “rectangular”, scale0 = 1, scaleN = 0.01, scaleCooling = “linear” as parameters in the Rsomoclu.train function of the Rsomoclu package. globalBmus is used to identify genes that cluster together. The cluster(s) that contain the gene(s) of interest are returned.</p>
<ul>
<li><p>number of nodes per dimension</p></li>
<li><p>Gene of interest</p></li>
<li>
<p>Heatmap</p>
<p>A heat map of the identied genes is displayed. Cluster are not the SOM cluster (since all the genes are from the same cluster).</p>
</li>
<li>
<p>Identified gene list</p>
<p>For better copy/paste operations the list of genes is displayed as a comma separated list.</p>
</li>
</ul>
</div>
</div>
<div id="data-exploration" class="section level4">
<h4 class="hasAnchor">
<a href="#data-exploration" class="anchor"></a>Data Exploration</h4>
<div id="expression" class="section level5">
<h5 class="hasAnchor">
<a href="#expression" class="anchor"></a>Expression</h5>
<div class="figure" style="text-align: center">
<img src="images/dataexplorationExpression.png" alt="Data exploration - Expression" width="90%"><p class="caption">
Data exploration - Expression
</p>
</div>
<p>As a convenience the following three plots are combined in one view. Most of the functionality displayed here can also be achieved using other tabs/views.</p>
<ul>
<li>
<p>Enter gene</p>
<p>Gene name or comma separated list of genes whoes expression level is being used to color cells in the 3D plot and violin plot.</p>
</li>
<li>
<p>3D plot</p>
<p>tSNE plot of the first 3 tSNE projections is shown, colored by the expression of the genes of interest.</p>
</li>
<li>
<p>2D plot</p>
<p>Standard 2D plot. Selections have no influence anywhere.</p>
</li>
<li>
<p>Violin plot</p>
<p>Whereas the violin plot under Co-expression-Violin shows the number of cells the express (or not) a given gene, here, the expression is estimated. The x-axis is fixed to clusters.</p>
</li>
</ul>
</div>
<div id="panel-plot" class="section level5">
<h5 class="hasAnchor">
<a href="#panel-plot" class="anchor"></a>Panel plot</h5>
<div class="figure" style="text-align: center">
<img src="images/DataexplorationPanel.png" alt="Data exploration - Panel plot" width="90%"><p class="caption">
Data exploration - Panel plot
</p>
</div>
<p>2D plots of genes of interest. If the x-axis is a categorical value and the y-axis is UMI.counts the y-axis related to the count for that gene. Otherwise, all genes are used. This plots automatically creates box plots if the x-axis is a categorical value. If a categorical value is chosen for the y-axis an error message is displayed (‘min’ not meaningful for factors). Choosing barcodes as an axis can cause the computer to hang. When y-axis is set to UMI.counts the most useful values for the x-axis are sampleNames or dbCluster, which allows comparing the expression of different genes in parallel over different samples or clusters.</p>
<ul>
<li>
<p>Cluster</p>
<p>Choose either all clusters combined (All) or an individual cluster</p>
</li>
<li>
<p>X</p>
<p>Projection to be used for the x-axis.</p>
</li>
<li>
<p>Y</p>
<p>Projection to be used for the y-axis.</p>
</li>
<li>
<p>Comma separated gene names</p>
<p>List of genes for which a 2D plot should be gerenated.</p>
</li>
<li><p>Plots</p></li>
</ul>
</div>
</div>
<div id="subcluster-analysis" class="section level4">
<h4 class="hasAnchor">
<a href="#subcluster-analysis" class="anchor"></a>Subcluster analysis</h4>
<div id="dge-analysis" class="section level5">
<h5 class="hasAnchor">
<a href="#dge-analysis" class="anchor"></a>DGE analysis</h5>
<div class="figure" style="text-align: center">
<img src="images/subclusteranalysis1.png" alt="Subcluster analysis - DGE analysis before selection" width="90%"><p class="caption">
Subcluster analysis - DGE analysis before selection
</p>
</div>
<div class="figure" style="text-align: center">
<img src="images/subclusteranalysis2.png" alt="Subcluster analysis - DGE analysis" width="90%"><p class="caption">
Subcluster analysis - DGE analysis
</p>
</div>
<p>Differential gene expression analysis of selected genes.</p>
<ul>
<li>
<p>Cluster</p>
<p>Selection of clusters to be used/displayed</p>
</li>
<li>
<p>X</p>
<p>X coordinate for both plots</p>
</li>
<li>
<p>Y</p>
<p>Y coordinate for both plots</p>
</li>
<li>
<p>Plot 1</p>
<p>Plot for the selection of case samples</p>
</li>
<li>
<p>Plot 2</p>
<p>Plot for the selection of control samples</p>
</li>
<li>
<p>Method to use</p>
<ul>
<li>
<p>Chi-square test</p>
<p>A chi-square test is performed on an estimated binomial distribution. This is one of the only remaining parts from the original CellView app that was used to help get started with this application.</p>
</li>
<li>
<p>t-test</p>
<p>Standard t-test from the R-stats package is used to calculate the p-values.</p>
</li>
</ul>
</li>
<li>
<p>Table: differentially expressed genes</p>
<p>Table with the differentially expressed genes. If a gene is higher expressed control samples (right plot) the average difference will be negative.</p>
</li>
</ul>
</div>
</div>
</div>
<div id="reused-gui-elements-modules" class="section level3">
<h3 class="hasAnchor">
<a href="#reused-gui-elements-modules" class="anchor"></a>Reused GUI elements (Modules)</h3>
<div id="tables" class="section level4">
<h4 class="hasAnchor">
<a href="#tables" class="anchor"></a>Tables</h4>
<div class="figure" style="text-align: center">
<img src="images/table1.png" alt="Modular tables" width="90%"><p class="caption">
Modular tables
</p>
</div>
<div class="figure" style="text-align: center">
<img src="images/tableSelect2.png" alt="Modular tables - two rows selected" width="90%"><p class="caption">
Modular tables - two rows selected
</p>
</div>
<div class="figure" style="text-align: center">
<img src="images/tableReorder.png" alt="Modular tables - reorder" width="90%"><p class="caption">
Modular tables - reorder
</p>
</div>
<div class="figure" style="text-align: center">
<img src="images/tableAll.png" alt="Modular tables - all selected" width="90%"><p class="caption">
Modular tables - all selected
</p>
</div>
<p>A modular table can be downloaded using the “Download Table” button.</p>
<ul>
<li>
<p>Selected names to be copied</p>
<p>When selecting row the corresponding row names are displayed as a comma separated list. This is useful for copy/paste actions in different other views.</p>
</li>
<li>
<p>Select all rows</p>
<p>Can be used to select all rows of the table and display the row names, but also through repetetive selection/unselection the selection can be removed.</p>
</li>
<li>
<p>reorder cells by sum of selected genes</p>
<p>Only the first 20 columns are shown. Since it is not necessarily clear that the first columns are the most interesting the user has the option to select genes and the reorder the columns based on the sum of these genes per cell. This happens when this option is checked. This option is not always applicable (e.g. DGE analysis)</p>
</li>
<li>
<p>Table</p>
<p>The table itself can be sorted (rows) by clicking on the column header. The sort order is indicated with a small triangle. Columns can also be filtered and a full text search is possible. The number of displayed rows can be changed as well.</p>
</li>
</ul>
</div>
<div id="d-plots" class="section level4">
<h4 class="hasAnchor">
<a href="#d-plots" class="anchor"></a>2D plots</h4>
<div class="figure" style="text-align: center">
<img src="images/2Dplot1.png" alt="Modular 2D plots" width="90%"><p class="caption">
Modular 2D plots
</p>
</div>
<div class="figure" style="text-align: center">
<img src="images/2dplotSmallGrp.png" alt="Modular 2D plots" width="90%"><p class="caption">
Modular 2D plots
</p>
</div>
<div class="figure" style="text-align: center">
<img src="images/2dplotAddOptions.png" alt="Modular 2D plots - additional options" width="90%"><p class="caption">
Modular 2D plots - additional options
</p>
</div>
<p>Except for the Co-expression - Selected tab the selection of cells is not used.</p>
<ul>
<li>
<p>comma separated list of genes for UmiCountPerGenes (left / right)</p>
<p>A list of comma separated gene names. In the columns below the sum of the expression per cell is used. Using these two fields one can plot the summed expression of two groups of gene (or individual genes) to visualize the cells that co-express these genes.</p>
</li>
<li>
<p>X, Y, color</p>
<p>The fields allow the selection of the X and Y coordinate as well as selection of the color used for the cells.</p>
</li>
<li>
<p>2D plot</p>
<p>When hovering over the right top of the figure a collection of options for saving the plot, changing the selection style and others will be made visible. Zooming is possible in different ways (mouse, top right menu)</p>
</li>
<li>
<p>Show more options</p>
<p>Upon selection other parameters are made visible.</p>
<ul>
<li><p>Log transform X</p></li>
<li>
<p>Log transform Y</p>
<p>Whether or not to show the data in log scale. Log with base 10 is used.</p>
</li>
<li><p>Divide X by</p></li>
<li>
<p>Divide Y by</p>
<p>Divide the values of the X/Y axis with the values from this projection. This is useful for identifying dead cells, when using the “before.filter” projection or setting the UMICountsPerGene to ribosomal proteins (as a comma separated list)</p>
</li>
<li>
<p>Add to group/otherwise overwrite</p>
<p>When checked, clicking the “change current selection” button will add the selected cells from the plot to the named group</p>
</li>
<li>
<p>Name group</p>
<p>Here the name of a group can be changed. This will generate a new group. It is not possible to remove a group.</p>
</li>
<li>
<p>Group names</p>
<p>A drop box selection for a group. This will also change the selection in the plot. This allows one to visualize a previously defined group of cells. Unfortunately it is not possible</p>
</li>
<li>
<p>Number of visible cells in section</p>
<p>Indicates the number of cells in a given group</p>
</li>
<li>
<p>Change current selection</p>
<p>Using this button one writes the changes to memory</p>
</li>
</ul>
</li>
</ul>
<p>Comments on groups:</p>
<p>Handling the groups is relatively complex because of the different possibilities especially with selection process within the plot. Once the group is activated (either after selecting a group in the group names dropdown or by clicking on “change current selection” the selected cells are colored red (instead of highlighted when selecting using the box/lasso select). This selection overwrites the selection in the plot. To deactivate this behaviour one has to uncheck the “show more options” check-box. Unfortunately the selection box/highlighting is removed in the plot though the selection is still used in the corresponding visualizations. This can be confusing.</p>
</div>
<div id="heatmaps" class="section level4">
<h4 class="hasAnchor">
<a href="#heatmaps" class="anchor"></a>Heatmaps</h4>
<div class="figure" style="text-align: center">
<img src="images/heatmap1.png" alt="Modular heatmaps" width="90%"><p class="caption">
Modular heatmaps
</p>
</div>
<div class="figure" style="text-align: center">
<img src="images/heatmap2.png" alt="Modular heatmaps - order cells" width="90%"><p class="caption">
Modular heatmaps - order cells
</p>
</div>
<div class="figure" style="text-align: center">
<img src="images/heatmap3.png" alt="Modular heatmaps -  dendrogram" width="90%"><p class="caption">
Modular heatmaps - dendrogram
</p>
</div>
<p>Like most figures the heatmap is also resizable using the lower right corner that can be dragged.</p>
<p>The color of the samples and clusters can be changed under Parameters - general parameters.</p>
<ul>
<li>
<p>Show more options</p>
<p>Once selected additional options for sorting and displaying annotion appear.</p>
<ul>
<li>
<p>Show tree for cells</p>
<p>The check-box overwrites the order (if any) that is described below. The columns are odered corresponding to the hirachical clustering on the cells.</p>
</li>
<li>
<p>Group names</p>
<p>Here, different projections can be chosen that appear as column annotations</p>
</li>
<li>
<p>Order of columns</p>
<p>The columns can be ordered manually by any of the projections available.</p>
</li>
</ul>
</li>
</ul>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Bernd Jagla.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
