<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Follow scran workflow to analyze single-cell RNA-seq data • SCHNAPPs</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><meta property="og:title" content="Follow scran workflow to analyze single-cell RNA-seq data">
<meta property="og:description" content="SCHNAPPs">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">SCHNAPPs</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.4.55</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/Readme.html">Readme</a>
    </li>
    <li>
      <a href="../../articles/pkdown/FAQs.html">Frequently asked questions for common tasks</a>
    </li>
    <li>
      <a href="../../articles/pkdown/SCHNAPPs_usage.html">SCHNAPPs - Single Cell sHiNy APP(s)</a>
    </li>
    <li>
      <a href="../../articles/pkdown/scranWorkflow.html">Follow scran workflow to analyze single-cell RNA-seq data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/C3BI-pasteur-fr/UTechSCB-SCHNAPPs/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="scranWorkflow_files/header-attrs-2.2/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Follow scran workflow to analyze single-cell RNA-seq data</h1>
                        <h4 class="author">Bernd Jagla</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:bernd.jagla@gmail.com" class="email">bernd.jagla@gmail.com</a>
      
                  
            <h4 class="date">2.6.2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/C3BI-pasteur-fr/UTechSCB-SCHNAPPs/blob/master/vignettes/pkdown/scranWorkflow.Rmd"><code>vignettes/pkdown/scranWorkflow.Rmd</code></a></small>
      <div class="hidden name"><code>scranWorkflow.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>This vignette shows how SCHNAPPs can be applied as a single-cell RNA-sequencing platform by comparing it to a standard workflow as decribed under SCRAN (<a href="https://bioconductor.org/packages/release/bioc/vignettes/scran/inst/doc/scran.html" class="uri">https://bioconductor.org/packages/release/bioc/vignettes/scran/inst/doc/scran.html</a>).</p>
</div>
<div id="data-setup" class="section level1">
<h1 class="hasAnchor">
<a href="#data-setup" class="anchor"></a>Data setup</h1>
<p>We pre-compute the scran workflow and make its results available for comparison within SCHNAPPs.</p>
<p>We use the same data set. Here, we have to create a singleCellExperiment object and save it in an external file to be able to load into SCHNAPPs later.</p>
<p>We prepare the singleCellExperiment object to contain the col/row Data that is needed by SCHNAPPs. At this point, which is usually performed by the bioinformatician who is preparing the data, we can also add other information and documentation.</p>
<p>colData has to contain the columns barcode = a unique identifier per experiment sampleNames = a name for a collection of cells.</p>
<p>rowData has to contian the columns symbol = the gene symbol Description = a text describing the gene id = unique identifier, e.g. ENSG number</p>
<p>The following stats</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="no">qcstats</span> <span class="kw">&lt;-</span> <span class="kw pkg">scater</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/perCellQCMetrics.html">perCellQCMetrics</a></span>(<span class="no">sce</span>)
<span class="fu">colData</span>(<span class="no">sce</span>) <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(<span class="fu">colData</span>(<span class="no">sce</span>), <span class="no">qcstats</span>)
<span class="fu">colData</span>(<span class="no">sce</span>)$<span class="no">barcode</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="fu">colData</span>(<span class="no">sce</span>))
<span class="fu">colData</span>(<span class="no">sce</span>)$<span class="no">sampleNames</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="fu">colData</span>(<span class="no">sce</span>)$<span class="no">sample</span>)

<span class="fu">rowData</span>(<span class="no">sce</span>)$<span class="no">Description</span> <span class="kw">=</span> <span class="st">""</span>
<span class="fu">rowData</span>(<span class="no">sce</span>)$<span class="no">id</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="fu">rowData</span>(<span class="no">sce</span>))</pre></body></html></div>
<p>To show the effect of the scater filtering we also calculate those. Missing values are handled as</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="no">qcfilter</span> <span class="kw">&lt;-</span> <span class="fu">quickPerCellQC</span>(<span class="no">qcstats</span>, <span class="kw">percent_subsets</span><span class="kw">=</span><span class="st">"altexps_ERCC_percent"</span>)
<span class="no">qcfilter</span>[<span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="no">qcfilter</span>$<span class="no">high_altexps_ERCC_percent</span>),<span class="st">"high_altexps_ERCC_percent"</span>] <span class="kw">=</span> <span class="no">T</span>
<span class="fu">colData</span>(<span class="no">sce</span>) <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(<span class="fu">colData</span>(<span class="no">sce</span>), <span class="no">qcfilter</span>)</pre></body></html></div>
<p>Now start SCHNAPPs.</p>
<p>We reset the standards to empty for the gene and cell filtering to not introduce any bias</p>
<p>Gene selection - regular expression for selection of genes to be removed = "" Gene selection - Min expression over all cell = 1 Cell selection - Min # of UMIs = 1</p>
<p>and then load the data we just created.</p>
<p>We now show which filters have been applied to the data in the scran workflow. We use the perCellQCMetrics metrics that cannot be calculated from within SCHNAPPs but could be precalculated. We can then visualize the cells that are filtered out and recover the thresholds used in individual steps:</p>
<ol style="list-style-type: decimal">
<li>based on ERCC percentage:</li>
</ol>
<p>This percentage is precalculated from the scr</p>
<p>The data can be visualized in the 2D plot (co-expression - selected) with the following paramters: X : altexps_ERCC_percent Y: sample color: high_altexps_ERCC_Percent</p>
<div class="figure" style="text-align: center">
<img src="images/scran_1.png" alt="Filter based on ERCC percentage" width="90%"><p class="caption">
Filter based on ERCC percentage
</p>
</div>
<ol start="2" style="list-style-type: decimal">
<li>filter low expressing cells</li>
</ol>
<p>Using a similar 2D plot with the following paramters: X : detected Y: sample color: low_n_features</p>
<p>we can selecte and remove cells with low expressing features.</p>
<div class="figure" style="text-align: center">
<img src="images/scran_2.png" alt="Filter based on detected genes" width="90%"><p class="caption">
Filter based on detected genes
</p>
</div>
<p>We end up with the same 1219 cells as in the vignette from scran.</p>
</div>
<div id="normalization" class="section level1">
<h1 class="hasAnchor">
<a href="#normalization" class="anchor"></a>Normalization</h1>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="co"># here we redo the scran workflow with different clustering algorithms </span>
<span class="co"># and store the results in the singleCellExperiment object that we are going to use in SCHNAPPs</span>

<span class="no">sce</span> <span class="kw">&lt;-</span> <span class="no">sce</span>[,!<span class="no">qcfilter</span>$<span class="no">discard</span>]
<span class="co"># summary(qcfilter$discard)</span>

<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">scran</span>)
<span class="no">clusters</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/quickCluster.html">quickCluster</a></span>(<span class="no">sce</span>)
<span class="no">sce</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/computeSumFactors.html">computeSumFactors</a></span>(<span class="no">sce</span>, <span class="kw">clusters</span><span class="kw">=</span><span class="no">clusters</span>)
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="fu">sizeFactors</span>(<span class="no">sce</span>))</pre></body></html></div>
<pre><code>##     Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
## 0.006722 0.442629 0.801312 1.000000 1.295809 9.227575</code></pre>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="no">sce</span> <span class="kw">&lt;-</span> <span class="fu">logNormCounts</span>(<span class="no">sce</span>)

<span class="no">dec</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html">modelGeneVar</a></span>(<span class="no">sce</span>)
<span class="co"># plot(dec$mean, dec$total, xlab="Mean log-expression", ylab="Variance")</span>
<span class="co"># curve(metadata(dec)$trend(x), col="blue", add=TRUE)</span>

<span class="no">dec2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVarWithSpikes.html">modelGeneVarWithSpikes</a></span>(<span class="no">sce</span>, <span class="st">'ERCC'</span>)

<span class="co"># Get the top 10% of genes.</span>
<span class="no">top.hvgs</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html">getTopHVGs</a></span>(<span class="no">dec</span>, <span class="kw">prop</span><span class="kw">=</span><span class="fl">0.1</span>)
<span class="no">sce</span> <span class="kw">&lt;-</span> <span class="fu">runPCA</span>(<span class="no">sce</span>, <span class="kw">subset_row</span><span class="kw">=</span><span class="no">top.hvgs</span>)
<span class="co"># reducedDimNames(sce)</span>
<span class="no">sced</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/denoisePCA.html">denoisePCA</a></span>(<span class="no">sce</span>, <span class="no">dec2</span>, <span class="kw">subset.row</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html">getTopHVGs</a></span>(<span class="no">dec2</span>, <span class="kw">prop</span><span class="kw">=</span><span class="fl">0.1</span>))
<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="fu">reducedDim</span>(<span class="no">sced</span>, <span class="st">"PCA"</span>))</pre></body></html></div>
<pre><code>## [1] 50</code></pre>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="no">output</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getClusteredPCs.html">getClusteredPCs</a></span>(<span class="fu">reducedDim</span>(<span class="no">sce</span>))
<span class="no">npcs</span> <span class="kw">&lt;-</span> <span class="fu">metadata</span>(<span class="no">output</span>)$<span class="no">chosen</span>
<span class="fu">reducedDim</span>(<span class="no">sce</span>, <span class="st">"PCAsub"</span>) <span class="kw">&lt;-</span> <span class="fu">reducedDim</span>(<span class="no">sce</span>, <span class="st">"PCA"</span>)[,<span class="fl">1</span>:<span class="no">npcs</span>,<span class="kw">drop</span><span class="kw">=</span><span class="fl">FALSE</span>]
<span class="co"># npcs</span>
<span class="co"># In this case, using the PCs that we chose from getClusteredPCs().</span>
<span class="no">g1</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/buildSNNGraph.html">buildSNNGraph</a></span>(<span class="no">sce</span>, <span class="kw">use.dimred</span><span class="kw">=</span><span class="st">"PCAsub"</span>)
<span class="no">cluster</span> <span class="kw">&lt;-</span> <span class="kw pkg">igraph</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/cluster_walktrap.html">cluster_walktrap</a></span>(<span class="no">g1</span>)$<span class="no">membership</span>
<span class="no">sce</span>$<span class="no">cluster</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="no">cluster</span>)

<span class="co"># In this case, using the PCs that we chose from getClusteredPCs().</span>
<span class="no">g</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/buildSNNGraph.html">buildSNNGraph</a></span>(<span class="no">sce</span>, <span class="kw">use.dimred</span><span class="kw">=</span><span class="st">"PCAsub"</span>, <span class="kw">type</span> <span class="kw">=</span> <span class="st">"jaccard"</span>)
<span class="no">cluster</span> <span class="kw">&lt;-</span> <span class="kw pkg">igraph</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/cluster_walktrap.html">cluster_walktrap</a></span>(<span class="no">g</span>)$<span class="no">membership</span>
<span class="no">sce</span>$<span class="no">cluster2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="no">cluster</span>)

<span class="no">g</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/buildSNNGraph.html">buildSNNGraph</a></span>(<span class="no">sce</span>, <span class="kw">use.dimred</span><span class="kw">=</span><span class="st">"PCAsub"</span>, <span class="kw">type</span> <span class="kw">=</span> <span class="st">"number"</span>)
<span class="no">cluster</span> <span class="kw">&lt;-</span> <span class="kw pkg">igraph</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/cluster_walktrap.html">cluster_walktrap</a></span>(<span class="no">g</span>)$<span class="no">membership</span>
<span class="no">sce</span>$<span class="no">cluster3</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="no">cluster</span>)



<span class="co"># table(sce$cluster) </span>
<span class="no">sce</span> <span class="kw">&lt;-</span> <span class="fu">runTSNE</span>(<span class="no">sce</span>, <span class="kw">dimred</span><span class="kw">=</span><span class="st">"PCAsub"</span>)
<span class="co"># plotTSNE(sce, colour_by="cluster", text_by="cluster")</span>
<span class="co"># plotTSNE(sce, colour_by="cluster2", text_by="cluster2")</span>
<span class="co"># plotTSNE(sce, colour_by="cluster3", text_by="cluster3")</span>


<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">pheatmap</span>)
<span class="no">ass.prob</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/bootstrapCluster.html">bootstrapCluster</a></span>(<span class="no">sce</span>, <span class="kw">FUN</span><span class="kw">=</span><span class="kw">function</span>(<span class="no">x</span>) {
    <span class="no">g</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/buildSNNGraph.html">buildSNNGraph</a></span>(<span class="no">x</span>, <span class="kw">use.dimred</span><span class="kw">=</span><span class="st">"PCAsub"</span>)
    <span class="kw pkg">igraph</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/cluster_walktrap.html">cluster_walktrap</a></span>(<span class="no">g</span>)$<span class="no">membership</span>
}, <span class="kw">clusters</span><span class="kw">=</span><span class="no">sce</span>$<span class="no">cluster</span>)

<span class="co"># pheatmap(ass.prob, cluster_cols=FALSE, cluster_rows=FALSE,</span>
<span class="co">#     col=colorRampPalette(c("white", "blue"))(100))</span></pre></body></html></div>
<p>We save all results from to be able to compare to what SCHNAPPs has to offer</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/save.html">save</a></span>(<span class="kw">file</span> <span class="kw">=</span> <span class="st">"scran.grunPancreas.RData"</span>, <span class="kw">list</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"sce"</span>))</pre></body></html></div>
<div class="figure" style="text-align: center">
<img src="images/scran_3.png" alt="Cluster assignment in tSNE space based on SCHNAPPs quickCluster" width="90%"><p class="caption">
Cluster assignment in tSNE space based on SCHNAPPs quickCluster
</p>
</div>
<div class="figure" style="text-align: center">
<img src="images/scran_4.png" alt="Cluster assignment in tSNE space based on rank" width="90%"><p class="caption">
Cluster assignment in tSNE space based on rank
</p>
</div>
<div class="figure" style="text-align: center">
<img src="images/scran_5.png" alt="Cluster assignment in tSNE space based on Jaccard" width="90%"><p class="caption">
Cluster assignment in tSNE space based on Jaccard
</p>
</div>
<div class="figure" style="text-align: center">
<img src="images/scran_6.png" alt="Cluster assignment in tSNE space based on number" width="90%"><p class="caption">
Cluster assignment in tSNE space based on number
</p>
</div>
<p>cluster = buildSNNGraph(type = “rank”)</p>
<div class="figure" style="text-align: center">
<img src="images/scran_7.png" alt="Alluvial plot comparision between quickcluster and rank" width="90%"><p class="caption">
Alluvial plot comparision between quickcluster and rank
</p>
</div>
<div class="figure" style="text-align: center">
<img src="images/scran_8.png" alt="Alluvial plot comparision between quickcluster and Jaccard" width="90%"><p class="caption">
Alluvial plot comparision between quickcluster and Jaccard
</p>
</div>
<div class="figure" style="text-align: center">
<img src="images/scran_9.png" alt="Alluvial plot comparision between quickcluster and number" width="90%"><p class="caption">
Alluvial plot comparision between quickcluster and number
</p>
</div>
<div class="figure" style="text-align: center">
<img src="images/scran_10.png" alt="Alluvial plot comparision between rank and Jaccard" width="90%"><p class="caption">
Alluvial plot comparision between rank and Jaccard
</p>
</div>
<div class="figure" style="text-align: center">
<img src="images/scran_11.png" alt="Alluvial plot comparision between rank and number" width="90%"><p class="caption">
Alluvial plot comparision between rank and number
</p>
</div>
<div class="figure" style="text-align: center">
<img src="images/scran_12.png" alt="heatmap of findMarker genes" width="90%"><p class="caption">
heatmap of findMarker genes
</p>
</div>
<div class="figure" style="text-align: center">
<img src="images/scran_13.png" alt="Heatmaps can be modified using additional options" width="90%"><p class="caption">
Heatmaps can be modified using additional options
</p>
</div>
<div class="figure" style="text-align: center">
<img src="images/scran_14.png" alt="Violin plot with permutataions" width="90%"><p class="caption">
Violin plot with permutataions
</p>
</div>
<div class="figure" style="text-align: center">
<img src="images/scran_15.png" alt="grouped violin plots" width="90%"><p class="caption">
grouped violin plots
</p>
</div>
<div class="figure" style="text-align: center">
<img src="images/scran_16.png" alt="3D tSNE plot with expression of specified genes" width="90%"><p class="caption">
3D tSNE plot with expression of specified genes
</p>
</div>
<div class="figure" style="text-align: center">
<img src="images/scran_17.png" alt="Panel Plot" width="90%"><p class="caption">
Panel Plot
</p>
</div>
<div class="figure" style="text-align: center">
<img src="images/scran_18.png" alt="DGE between specific groups of cells" width="90%"><p class="caption">
DGE between specific groups of cells
</p>
</div>
<div class="figure" style="text-align: center">
<img src="images/scran_19.png" alt="DGE results as Volcano plot" width="90%"><p class="caption">
DGE results as Volcano plot
</p>
</div>
<div class="figure" style="text-align: center">
<img src="images/scran_20.png" alt="Table of DGE" width="90%"><p class="caption">
Table of DGE
</p>
</div>
<p>SCHNAPPs uses the projection “dbCluster” for the current clustering result. In order to compare different clustering procedures the rename projection tool allows the user to copy the results do a different name.</p>
<div class="figure" style="text-align: center">
<img src="images/scran_21.png" alt="Projections can be renamed" width="90%"><p class="caption">
Projections can be renamed
</p>
</div>
<p>Cells can be grouped and named using the 2D plot, Those are are binary groups, i.e. a cells belongs either to the group or not. There are cases where this binary lababling is not sufficient. For example if there are multiple indipendant groups of cells. The combine projections tabs allows combining such groups. In addition it is also possible to rename levels of factorials. Those are also copied to a new projection such that no information is lost.</p>
<div class="figure" style="text-align: center">
<img src="images/scran_22.png" alt="Select cells for copy/paste" width="90%"><p class="caption">
Select cells for copy/paste
</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Bernd Jagla.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
