<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Follow scran workflow to analyze single-cell RNA-seq data • SCHNAPPs</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><meta property="og:title" content="Follow scran workflow to analyze single-cell RNA-seq data">
<meta property="og:description" content="SCHNAPPs">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">SCHNAPPs</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.12.73</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/Readme.html">Readme</a>
    </li>
    <li>
      <a href="../../articles/pkdown/FAQs.html">FAQs</a>
    </li>
    <li>
      <a href="../../articles/pkdown/SCHNAPPs_usage.html">SCHNAPPs - Single Cell sHiNy APP(s)</a>
    </li>
    <li>
      <a href="../../articles/pkdown/SeuratWorkflow.html">Seurat workflow</a>
    </li>
    <li>
      <a href="../../articles/pkdown/scranWorkflow.html">Follow scran workflow to analyze single-cell RNA-seq data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/C3BI-pasteur-fr/UTechSCB-SCHNAPPs/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><link href="scranWorkflow_files/vembedr-0.1.5/css/vembedr.css" rel="stylesheet">
<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Follow scran workflow to analyze single-cell
RNA-seq data</h1>
                        <h4 data-toc-skip class="author">Bernd
Jagla</h4>
                  <a class="author_email" href="mailto:#">bernd <dot>
jagla <at> pasteur <dot> fr</dot></at></dot></a>
      
                  
            <h4 data-toc-skip class="date">2.6.2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/C3BI-pasteur-fr/UTechSCB-SCHNAPPs/blob/HEAD/vignettes/pkdown/scranWorkflow.Rmd" class="external-link"><code>vignettes/pkdown/scranWorkflow.Rmd</code></a></small>
      <div class="hidden name"><code>scranWorkflow.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/C3BI-pasteur-fr/UTechSCB-SCHNAPPs" class="external-link">SCHNAPPs</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/register.html" class="external-link">register</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SerialParam-class.html" class="external-link">SerialParam</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="co"># avoid problems with fastMNN parallelization.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette shows how SCHNAPPs can be applied as a single-cell
RNA-sequencing platform by comparing it to a standard workflow as
decribed under SCRAN (<a href="https://bioconductor.org/packages/release/bioc/vignettes/scran/inst/doc/scran.html" class="external-link uri">https://bioconductor.org/packages/release/bioc/vignettes/scran/inst/doc/scran.html</a>).</p>
</div>
<div class="section level2">
<h2 id="data-setup">Data setup<a class="anchor" aria-label="anchor" href="#data-setup"></a>
</h2>
<p>We pre-compute the scran workflow and make its results available for
comparison within SCHNAPPs.</p>
<p>We use the same data set. Here, we have to create a
singleCellExperiment object and save it in an external file to be able
to load into SCHNAPPs later.</p>
<p>In case of error this might work: devtools::install_version(“dbplyr”,
version = “2.3.4”)</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span> <span class="st">"scRNAseq"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"scRNAseq"</span><span class="op">)</span><span class="op">}</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scRNAseq</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/GrunPancreasData.html" class="external-link">GrunPancreasData</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">sce</span></span>
<span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 20064 1728 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(20064): A1BG-AS1__chr19 A1BG__chr19 ... ZZEF1__chr17</span></span>
<span><span class="co">##   ZZZ3__chr1</span></span>
<span><span class="co">## rowData names(2): symbol chr</span></span>
<span><span class="co">## colnames(1728): D2ex_1 D2ex_2 ... D17TGFB_95 D17TGFB_96</span></span>
<span><span class="co">## colData names(2): donor sample</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: endogenous</span></span>
<span><span class="co">## altExpNames(1): ERCC</span></span></code></pre></div>
<p>We prepare the singleCellExperiment object to contain the col/row
Data that is needed by SCHNAPPs. At this point, which is usually
performed by the bioinformatician who is preparing the data, we can also
add other information and documentation.</p>
<p>colData has to contain the columns barcode = a unique identifier per
experiment sampleNames = a name for a collection of cells.</p>
<p>rowData has to contian the columns symbol = the gene symbol
Description = a text describing the gene id = unique identifier,
e.g. ENSG number</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qcstats</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/perCellQCMetrics.html" class="external-link">perCellQCMetrics</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="va">qcstats</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">barcode</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">sampleNames</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">Description</span> <span class="op">=</span> <span class="st">""</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">id</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>To show the effect of the scater filtering we also calculate those.
Missing values are handled as genes with high ERCC expressing genes.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qcfilter</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/quickPerCellQC.html" class="external-link">quickPerCellQC</a></span><span class="op">(</span><span class="va">qcstats</span>, percent_subsets<span class="op">=</span><span class="st">"altexps_ERCC_percent"</span><span class="op">)</span></span>
<span><span class="va">qcfilter</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">qcfilter</span><span class="op">$</span><span class="va">high_altexps_ERCC_percent</span><span class="op">)</span>,<span class="st">"high_altexps_ERCC_percent"</span><span class="op">]</span> <span class="op">=</span> <span class="cn">T</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="va">qcfilter</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># here we redo the scran workflow with different clustering algorithms </span></span>
<span><span class="co"># and store the results in the singleCellExperiment object that we are going to use in SCHNAPPs</span></span>
<span><span class="va">sceOrg</span> <span class="op">=</span> <span class="va">sce</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>,<span class="op">!</span><span class="va">qcfilter</span><span class="op">$</span><span class="va">discard</span><span class="op">]</span></span>
<span><span class="co"># summary(qcfilter$discard)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/quickCluster.html" class="external-link">quickCluster</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/computeSumFactors.html" class="external-link">computeSumFactors</a></span><span class="op">(</span><span class="va">sce</span>, clusters<span class="op">=</span><span class="va">clusters</span><span class="op">)</span></span>
<span><span class="co"># summary(sizeFactors(sce))</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="co"># plot(dec$mean, dec$total, xlab="Mean log-expression", ylab="Variance")</span></span>
<span><span class="co"># curve(metadata(dec)$trend(x), col="blue", add=TRUE)</span></span>
<span></span>
<span><span class="va">dec2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVarWithSpikes.html" class="external-link">modelGeneVarWithSpikes</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">'ERCC'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get the top 10% of genes.</span></span>
<span><span class="va">top.hvgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">dec</span>, prop<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sce</span>, subset_row<span class="op">=</span><span class="va">top.hvgs</span><span class="op">)</span></span>
<span><span class="co"># reducedDimNames(sce)</span></span>
<span><span class="va">sced</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/denoisePCA.html" class="external-link">denoisePCA</a></span><span class="op">(</span><span class="va">sce</span>, <span class="va">dec2</span>, subset.row<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">dec2</span>, prop<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># ncol(reducedDim(sced, "PCA"))</span></span>
<span><span class="va">output</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getClusteredPCs.html" class="external-link">getClusteredPCs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">npcs</span> <span class="op">&lt;-</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">output</span><span class="op">)</span><span class="op">$</span><span class="va">chosen</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"PCAsub"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"PCA"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="va">npcs</span>,drop<span class="op">=</span><span class="cn">FALSE</span><span class="op">]</span></span>
<span><span class="co"># npcs</span></span>
<span><span class="co"># In this case, using the PCs that we chose from getClusteredPCs().</span></span>
<span><span class="va">g1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/buildSNNGraph.html" class="external-link">buildSNNGraph</a></span><span class="op">(</span><span class="va">sce</span>, use.dimred<span class="op">=</span><span class="st">"PCAsub"</span><span class="op">)</span></span>
<span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/cluster_walktrap.html" class="external-link">cluster_walktrap</a></span><span class="op">(</span><span class="va">g1</span><span class="op">)</span><span class="op">$</span><span class="va">membership</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span></span>
<span><span class="co"># table(sce$cluster)</span></span>
<span></span>
<span><span class="va">g1a</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/buildSNNGraph.html" class="external-link">buildSNNGraph</a></span><span class="op">(</span><span class="va">sce</span>, use.dimred<span class="op">=</span><span class="st">"PCAsub"</span><span class="op">)</span></span>
<span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/cluster_walktrap.html" class="external-link">cluster_walktrap</a></span><span class="op">(</span><span class="va">g1a</span><span class="op">)</span><span class="op">$</span><span class="va">membership</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">cluster1a</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># In this case, using the PCs that we chose from getClusteredPCs().</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/buildSNNGraph.html" class="external-link">buildSNNGraph</a></span><span class="op">(</span><span class="va">sce</span>, use.dimred<span class="op">=</span><span class="st">"PCAsub"</span>, type <span class="op">=</span> <span class="st">"jaccard"</span><span class="op">)</span></span>
<span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/cluster_walktrap.html" class="external-link">cluster_walktrap</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span><span class="op">$</span><span class="va">membership</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">cluster2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span></span>
<span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/buildSNNGraph.html" class="external-link">buildSNNGraph</a></span><span class="op">(</span><span class="va">sce</span>, use.dimred<span class="op">=</span><span class="st">"PCAsub"</span>, type <span class="op">=</span> <span class="st">"number"</span><span class="op">)</span></span>
<span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/cluster_walktrap.html" class="external-link">cluster_walktrap</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span><span class="op">$</span><span class="va">membership</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">cluster3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span></span>
<span></span>
<span></span>
<span></span>
<span><span class="co"># table(sce$cluster) </span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html" class="external-link">runTSNE</a></span><span class="op">(</span><span class="va">sce</span>, dimred<span class="op">=</span><span class="st">"PCAsub"</span><span class="op">)</span></span>
<span><span class="co"># plotTSNE(sce, colour_by="cluster", text_by="cluster")</span></span>
<span><span class="co"># plotTSNE(sce, colour_by="cluster2", text_by="cluster2")</span></span>
<span><span class="co"># plotTSNE(sce, colour_by="cluster3", text_by="cluster3")</span></span>
<span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span></span>
<span><span class="va">ass.prob</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/defunct.html" class="external-link">bootstrapCluster</a></span><span class="op">(</span><span class="va">sce</span>, FUN<span class="op">=</span><span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/buildSNNGraph.html" class="external-link">buildSNNGraph</a></span><span class="op">(</span><span class="va">x</span>, use.dimred<span class="op">=</span><span class="st">"PCAsub"</span><span class="op">)</span></span>
<span>  <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/cluster_walktrap.html" class="external-link">cluster_walktrap</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span><span class="op">$</span><span class="va">membership</span></span>
<span><span class="op">}</span>, clusters<span class="op">=</span><span class="va">sce</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sceOrg</span><span class="op">$</span><span class="va">cluster</span> <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="va">sceOrg</span><span class="op">$</span><span class="va">cluster1a</span> <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="va">sceOrg</span><span class="op">$</span><span class="va">cluster2</span> <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="va">sceOrg</span><span class="op">$</span><span class="va">cluster3</span> <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">sceOrg</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sceOrg</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">sceOrg</span><span class="op">$</span><span class="va">cluster1a</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sceOrg</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">sceOrg</span><span class="op">$</span><span class="va">cluster2</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sceOrg</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">sceOrg</span><span class="op">$</span><span class="va">cluster3</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sceOrg</span><span class="op">)</span></span>
<span><span class="va">sceOrg</span><span class="op">$</span><span class="va">cluster</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">]</span> <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">cluster</span></span>
<span><span class="va">sceOrg</span><span class="op">$</span><span class="va">cluster1a</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">]</span> <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">cluster1a</span></span>
<span><span class="va">sceOrg</span><span class="op">$</span><span class="va">cluster2</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">]</span> <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">cluster2</span></span>
<span><span class="va">sceOrg</span><span class="op">$</span><span class="va">cluster3</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">]</span> <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">cluster3</span></span>
<span><span class="va">sceOrg</span><span class="op">$</span><span class="va">cluster</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">sceOrg</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span></span>
<span><span class="va">sceOrg</span><span class="op">$</span><span class="va">cluster1a</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">sceOrg</span><span class="op">$</span><span class="va">cluster1a</span><span class="op">)</span></span>
<span><span class="va">sceOrg</span><span class="op">$</span><span class="va">cluster2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">sceOrg</span><span class="op">$</span><span class="va">cluster2</span><span class="op">)</span></span>
<span><span class="va">sceOrg</span><span class="op">$</span><span class="va">cluster3</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">sceOrg</span><span class="op">$</span><span class="va">cluster3</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># pheatmap(ass.prob, cluster_cols=FALSE, cluster_rows=FALSE,</span></span>
<span><span class="co">#     col=colorRampPalette(c("white", "blue"))(100))</span></span></code></pre></div>
<p>We save all results including the cluster assignments from the scran
workflow to be able to compare to SCHNAPPs.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"scran.grunPancreas.RData"</span>, list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sceOrg"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="schnapps">SCHNAPPs<a class="anchor" aria-label="anchor" href="#schnapps"></a>
</h2>
<p>Now start SCHNAPPs.</p>
<p>Please see the help and other documentation on how to enable tracking
data changes and saving plots (historyPath option). The defaultValues
allows us to preset any user defineable variable in the GUI.</p>
<p>This is partially done to simplify your life as you don’t have to set
all the parameters partially to avoid mistakes and for this to show the
correct set-up.</p>
<p>The variables are described in the code chunk. To note: the number of
genes to use is set to 951, which is the same value as in the scran
vignette. We don’t have a way to calculate this. But we are aware that
changing this value has substaintial effects for the cluster assignment.
We strongly believe that the optimization of these parameters is not
trivial and we are currently looking for solutions how to solve this.
Similar argumenation goes for the number of PCs to use for the
clustering algorithm.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/C3BI-pasteur-fr/UTechSCB-SCHNAPPs" class="external-link">SCHNAPPs</a></span><span class="op">)</span></span>
<span><span class="va">defaultValues</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># input</span></span>
<span><span class="va">defaultValues</span><span class="op">[[</span><span class="st">"sampleInput"</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span><span class="cn">FALSE</span> <span class="co"># don't subsample</span></span>
<span><span class="va">defaultValues</span><span class="op">[[</span><span class="st">"whichscLog"</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="st">"calcLog"</span> <span class="co"># calculate normalization</span></span>
<span></span>
<span><span class="co"># Normaliztion</span></span>
<span><span class="va">defaultValues</span><span class="op">[[</span><span class="st">"normalizationRadioButton"</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="st">"DE_scaterNormalization"</span> <span class="co"># use scaterNorm</span></span>
<span></span>
<span><span class="co"># General parameters</span></span>
<span><span class="co">## PArameters for PCA</span></span>
<span><span class="va">defaultValues</span><span class="op">[[</span><span class="st">"pcaScale"</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="cn">FALSE</span> <span class="co"># don't scale</span></span>
<span><span class="va">defaultValues</span><span class="op">[[</span><span class="st">"pcaN"</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="fl">951</span> <span class="co"># number of genes to use.</span></span>
<span><span class="va">defaultValues</span><span class="op">[[</span><span class="st">"pcaRank"</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="fl">14</span> <span class="co"># number PCs to use for clustering</span></span>
<span></span>
<span></span>
<span><span class="co"># Parameters for clustering</span></span>
<span><span class="va">defaultValues</span><span class="op">[[</span><span class="st">"tabsetCluster"</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="st">"snnGraph"</span> <span class="co"># use SNN graph function</span></span>
<span></span>
<span><span class="co"># preset Alluvial plot axes</span></span>
<span><span class="va">defaultValues</span><span class="op">[[</span><span class="st">"alluiv1"</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="st">"cluster"</span></span>
<span><span class="va">defaultValues</span><span class="op">[[</span><span class="st">"alluiv2"</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="st">"dbCluster"</span></span>
<span></span>
<span></span>
<span><span class="co"># Gene selection</span></span>
<span><span class="co"># we usually remove genes from mitochrondrium and Ribosome associated proteins using the following regular expression ("^MT-|^RP|^MRP"). This is not done in the scran vignette </span></span>
<span><span class="va">defaultValues</span><span class="op">[[</span><span class="st">"selectIds"</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="st">""</span></span>
<span><span class="va">defaultValues</span><span class="op">[[</span><span class="st">"minGenes"</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="fl">1</span> <span class="co"># min number of reads per cell</span></span>
<span></span>
<span></span>
<span></span>
<span><span class="co"># Cell selection</span></span>
<span><span class="va">defaultValues</span><span class="op">[[</span><span class="st">"minGenesGS"</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="fl">1</span> <span class="co"># minimum number of UMIs (sum) over all cells.</span></span>
<span><span class="co"># cells to be removed</span></span>
<span><span class="va">defaultValues</span><span class="op">[[</span><span class="st">"cellsFiltersOut"</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="st">"D2ex_54, D2ex_64, D2ex_93, D2ex_95, D2ex_96, D3en1_8, D3en1_9, D3en1_12, D3en1_14, D3en1_15, D3en1_18, D3en1_19, D3en1_20, D3en1_21, D3en1_22, D3en1_23, D3en1_25, D3en1_31, D3en1_32, D3en1_34, D3en1_35, D3en1_36, D3en1_39, D3en1_43, D3en1_44, D3en1_47, D3en1_51, D3en1_52, D3en1_53, D3en1_57, D3en1_58, D3en1_59, D3en1_61, D3en1_62, D3en1_64, D3en1_66, D3en1_70, D3en1_72, D3en1_74, D3en1_76, D3en1_77, D3en1_78, D3en1_80, D3en1_82, D3en1_86, D3en1_91, D3en1_92, D3en1_95, D3en1_96, D3en2_1, D3en2_4, D3en2_8, D3en2_9, D3en2_10, D3en2_11, D3en2_14, D3en2_15, D3en2_18, D3en2_19, D3en2_20, D3en2_28, D3en2_29, D3en2_34, D3en2_35, D3en2_36, D3en2_40, D3en2_41, D3en2_43, D3en2_44, D3en2_46, D3en2_47, D3en2_49, D3en2_53, D3en2_54, D3en2_59, D3en2_65, D3en2_67, D3en2_69, D3en2_70, D3en2_75, D3en2_76, D3en2_77, D3en2_78, D3en2_79, D3en2_80, D3en2_82, D3en2_84, D3en2_86, D3en2_87, D3en2_88, D3en2_89, D3en2_95, D3en2_96, D3en3_3, D3en3_4, D3en3_12, D3en3_13, D3en3_14, D3en3_15, D3en3_17, D3en3_19, D3en3_20, D3en3_22, D3en3_23, D3en3_24, D3en3_25, D3en3_26, D3en3_31, D3en3_33, D3en3_35, D3en3_37, D3en3_38, D3en3_42, D3en3_43, D3en3_45, D3en3_47, D3en3_49, D3en3_56, D3en3_57, D3en3_58, D3en3_60, D3en3_61, D3en3_62, D3en3_63, D3en3_66, D3en3_69, D3en3_70, D3en3_71, D3en3_76, D3en3_79, D3en3_88, D3en3_90, D3en3_91, D3en3_93, D3en3_96, D3en4_1, D3en4_5, D3en4_10, D3en4_11, D3en4_13, D3en4_15, D3en4_17, D3en4_18, D3en4_21, D3en4_24, D3en4_27, D3en4_28, D3en4_29, D3en4_34, D3en4_36, D3en4_37, D3en4_40, D3en4_44, D3en4_45, D3en4_49, D3en4_50, D3en4_51, D3en4_52, D3en4_53, D3en4_55, D3en4_57, D3en4_59, D3en4_62, D3en4_66, D3en4_67, D3en4_70, D3en4_75, D3en4_76, D3en4_77, D3en4_78, D3en4_79, D3en4_80, D3en4_88, D3en4_93, D3en4_94, D3en4_95, D3ex_1, D3ex_2, D3ex_7, D3ex_8, D3ex_10, D3ex_11, D3ex_13, D3ex_14, D3ex_15, D3ex_17, D3ex_18, D3ex_19, D3ex_20, D3ex_21, D3ex_22, D3ex_23, D3ex_25, D3ex_26, D3ex_27, D3ex_28, D3ex_29, D3ex_30, D3ex_31, D3ex_32, D3ex_33, D3ex_35, D3ex_37, D3ex_38, D3ex_42, D3ex_43, D3ex_44, D3ex_45, D3ex_46, D3ex_47, D3ex_49, D3ex_50, D3ex_52, D3ex_53, D3ex_56, D3ex_57, D3ex_58, D3ex_59, D3ex_60, D3ex_61, D3ex_62, D3ex_63, D3ex_64, D3ex_65, D3ex_67, D3ex_69, D3ex_70, D3ex_72, D3ex_73, D3ex_75, D3ex_78, D3ex_79, D3ex_80, D3ex_81, D3ex_82, D3ex_84, D3ex_85, D3ex_86, D3ex_89, D3ex_93, D3ex_94, D3ex_95, D3ex_96, D74_4, D74_11, D74_19, D74_30, D74_38, D74_50, D74_72, D74_80, D71_13, D71_24, D71_25, D71_46, D71_50, D71_74, D71_77, D71_96, D72_6, D72_17, D72_26, D72_51, D72_56, D72_60, D73_53, D73_66, D73_74, D73_75, D73_77, D73_80, D73_88, D10631_1, D10631_6, D10631_13, D10631_16, D10631_28, D10631_30, D10631_31, D10631_32, D10631_35, D10631_49, D10631_59, D10631_77, D10631_86, D101_2, D101_3, D101_4, D101_6, D101_9, D101_12, D101_15, D101_18, D101_19, D101_20, D101_23, D101_26, D101_30, D101_32, D101_33, D101_38, D101_39, D101_40, D101_41, D101_47, D101_49, D101_54, D101_55, D101_60, D101_61, D101_62, D101_63, D101_64, D101_65, D101_67, D101_68, D101_70, D101_71, D101_75, D101_76, D101_78, D101_79, D101_80, D101_85, D101_86, D101_87, D101_90, D101_92, D101_94, D101_96, D102_2, D102_5, D102_8, D102_9, D102_10, D102_12, D102_13, D102_15, D102_16, D102_19, D102_20, D102_22, D102_23, D102_27, D102_28, D102_29, D102_30, D102_31, D102_32, D102_33, D102_34, D102_35, D102_36, D102_37, D102_38, D102_41, D102_42, D102_43, D102_44, D102_46, D102_48, D102_49, D102_52, D102_53, D102_56, D102_60, D102_62, D102_65, D102_67, D102_70, D102_72, D102_74, D102_77, D102_84, D102_85, D102_88, D102_89, D102_90, D102_92, D102_95, D17All1_13, D17All1_94, D17All2_5, D17All2_50, D17All2_89, D17TGFB_60, D17TGFB_62, D2ex_94, D3en1_40, D3en2_51, D3en3_18, D3en3_40, D3en3_77, D3en3_81, D3en3_83, D3en4_90, D3en4_96, D74_48, D74_84, D74_96, D71_89, D72_94, D73_47, D10631_47, D10631_75, D10631_78, D10631_79, D10631_80, D10631_83, D10631_84, D10631_94, D101_8, D101_11, D101_46, D102_47, D102_80, D102_81, D102_83, D172444_94, D172444_95, D172444_96, D17All1_14, D17All1_15, D17All1_17, D17All1_85, D17All2_96, D17TGFB_96"</span></span>
<span></span>
<span><span class="fu"><a href="../../reference/schnapps.html">schnapps</a></span><span class="op">(</span>defaultValues <span class="op">=</span> <span class="va">defaultValues</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="removal-of-cells">Removal of cells<a class="anchor" aria-label="anchor" href="#removal-of-cells"></a>
</h3>
<p>This section shows how to remove the cells that have been identified
in the scran vignette. To follow we have to remove the cells already
selected. and set “Cells to be removed” under Cell selection -
additional parameters to an empty string as shown below:</p>
<div class="figure" style="text-align: center">
<img src="images/scran_1a.png" alt="Filter based on ERCC percentage" width="90%"><p class="caption">
Filter based on ERCC percentage
</p>
</div>
<p>and click on apply changes.</p>
<p>Then load the data we just created.</p>
<div class="figure" style="text-align: center">
<img src="images/scran_1b.png" alt="Filter based on ERCC percentage" width="90%"><p class="caption">
Filter based on ERCC percentage
</p>
</div>
<p>We now show which filters have been applied to the data in the scran
workflow. We use the perCellQCMetrics metrics that cannot be calculated
from within SCHNAPPs but could be precalculated. We can then visualize
the cells that are filtered out and recover the thresholds used in
individual steps:</p>
<div class="section level4">
<h4 id="based-on-ercc-percentage">based on ERCC percentage:<a class="anchor" aria-label="anchor" href="#based-on-ercc-percentage"></a>
</h4>
<p>This percentage is precalculated from the scr</p>
<p>The data can be visualized in the 2D plot (co-expression - selected)
with the following paramters: X : altexps_ERCC_percent Y: sample color:
high_altexps_ERCC_Percent</p>
<div class="figure" style="text-align: center">
<img src="images/scran_1.png" alt="Filter based on ERCC percentage" width="90%"><p class="caption">
Filter based on ERCC percentage
</p>
</div>
<p>The selected cell names can be copy/pasted to the “Cells to be
removed” under Cell selection - additional parameters” as described
above.</p>
</div>
<div class="section level4">
<h4 id="filter-low-expressing-cells">filter low expressing cells<a class="anchor" aria-label="anchor" href="#filter-low-expressing-cells"></a>
</h4>
<p>Using a similar 2D plot with the following paramters: X : detected Y:
sample color: low_n_features</p>
<p>we can select and remove cells with low expressing features as
described above.</p>
<div class="figure" style="text-align: center">
<img src="images/scran_2.png" alt="Filter based on detected genes" width="90%"><p class="caption">
Filter based on detected genes
</p>
</div>
<p>We end up with the same 1219 cells as in the vignette from scran.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="normalization-clustering">Normalization / clustering<a class="anchor" aria-label="anchor" href="#normalization-clustering"></a>
</h2>
<p>We use “scaterNorm”, which is applying exactly the same procedure
that is used in the vignette. The user has several other options
available that are better described in the online documentation.</p>
<p>The snnGraph method is used for the clustering.</p>
</div>
<div class="section level2">
<h2 id="results-are-the-same">Results are the same<a class="anchor" aria-label="anchor" href="#results-are-the-same"></a>
</h2>
<p>The alluvial plot between the cluster assignment from the scran
package and the cluster assignment from SCHNAPPs (dbCluster) shows that
they are the same.</p>
<div class="figure" style="text-align: center">
<img src="images/alluvialAllGood.png" alt="Cluster assignment comparison using Alluvial plot" width="90%"><p class="caption">
Cluster assignment comparison using Alluvial plot
</p>
</div>
<p>The same result can be concluded from plotting the cluster
assignments in a 2D plot:</p>
<div class="figure" style="text-align: center">
<img src="images/2dAllGood.png" alt="Cluster assignment comparison using 2D plot" width="90%"><p class="caption">
Cluster assignment comparison using 2D plot
</p>
</div>
</div>
<div class="section level2">
<h2 id="characterization-of-results">Characterization of results<a class="anchor" aria-label="anchor" href="#characterization-of-results"></a>
</h2>
<p>The following will show a few possible results that can be achieved
using SCHNAPPs:</p>
<div class="section level3">
<h3 id="co-expression---all-clusters">Co-expression - All clusters<a class="anchor" aria-label="anchor" href="#co-expression---all-clusters"></a>
</h3>
<div class="figure" style="text-align: center">
<img src="images/heatmapAllCells.png" alt="Co-expression - All clusters; After list of gene names has been set to empty SCHNAPPs will automatically fill the list with characteristic genes for each cluster." width="90%"><p class="caption">
Co-expression - All clusters; After list of gene names has been set to
empty SCHNAPPs will automatically fill the list with characteristic
genes for each cluster.
</p>
</div>
<p>The following video shows how to use the heatmap module. If the field
with the gene names is empty the function FindMarkers is used to
identify the genes that are most characteristic for each cluster. This
list of genes is used for the heatmap.</p>
<div class="vembedr">
<div>
<iframe src="https://www.youtube.com/embed/aiKPccCmZgk" width="533" height="300" frameborder="0" allowfullscreen="" data-external="1"></iframe>
</div>
</div>
<p>The genes CPA1, CTRB1, PLA2G1B, PRSS1, PRSS3P2 are highly expressed
in cluster 2. GCG, TTR are highly expressed in cluster 5.</p>
</div>
<div class="section level3">
<h3 id="co-expression---selected">Co-expression - selected<a class="anchor" aria-label="anchor" href="#co-expression---selected"></a>
</h3>
<p>In the following we describe a few possibilities of the 2D plot</p>
<div class="section level4">
<h4 id="co-expression-of-two-lists-of-genes">co-expression of two lists of genes<a class="anchor" aria-label="anchor" href="#co-expression-of-two-lists-of-genes"></a>
</h4>
<p>We can now compare the genes previously found and see how these cells
express genes from different clusters.</p>
<div class="figure" style="text-align: center">
<img src="images/coexpresCl2Cl5.png" alt="2D plot of the sum of expression of two different lists of genes, colored by cluster association." width="90%"><p class="caption">
2D plot of the sum of expression of two different lists of genes,
colored by cluster association.
</p>
</div>
<p>The cell density can be used to identify where the majority of cells
lies or if and where many cells are drawn on top of eachother.</p>
<div class="figure" style="text-align: center">
<img src="images/2dCellDensity.png" alt="2D plot of the sum of expression of two different lists of genes, colored by density of cells." width="90%"><p class="caption">
2D plot of the sum of expression of two different lists of genes,
colored by density of cells.
</p>
</div>
<p>The composition of clusters relative to the donor information can be
also using the same tool as shown below:</p>
<div class="figure" style="text-align: center">
<img src="images/histogramDonor.png" alt="Histogram of cell count per cluster assignment and split by donor information." width="90%"><p class="caption">
Histogram of cell count per cluster assignment and split by donor
information.
</p>
</div>
<div class="figure" style="text-align: center">
<img src="images/nameGroupCell.png" alt="UMAP projection. Selecting and naming of a group of cells." width="90%"><p class="caption">
UMAP projection. Selecting and naming of a group of cells.
</p>
</div>
<p>Cells can be grouped and named using the 2D plot, Those are are
binary groups, i.e. a cells belongs either to the group or not.</p>
<p>The following video shows how to name groups of cells: * Cells have
to be selected in the 2D plot. * Give a name * ensure that the reference
for the selection is set to “plot” * click “change current
selection”</p>
<p>To note: * in-between, the cells seem not to be selected anymore, but
that is a “feature” of the underlying visualization functions. * using
the a previously defined group of cells allows to avoid selecting cells
that are potentially hidden. This can happen, when the 2D coordinates
have changed. This also allows a more cytometry way of thinking, where
concurrent selections of cells are used to filter.</p>
<p>After having selected multiple groups, we have these groups as
projections available with TRUE/FALSe assignment for each cell.
Sometimes, we want to combine these projections, e.g. if there are more
than two cell types. This is done under Parameters - Projections -
combine projections:</p>
<ul>
<li>Select the two projections that should be combined</li>
<li>name the new projection</li>
<li>repeat as needed.</li>
</ul>
<p>Since it is not ensured that these projections are not overlapping,
the individual values are concatenated and a new factor is created. The
levels of these factors can be changed under the “rename levels” tab.
The tables allows verifying the meaning of the levels.</p>
<p>When renaming the levels, don’t forget to add a meaningfull name,
otherwise “X” is chosen for you.</p>
<p>If this happens, don’t worry, this can be changed in the “rename
projections” tab, where projections can be deleted, but only those you
have created manually.</p>
<div class="vembedr">
<div>
<iframe src="https://www.youtube.com/embed/xOSWcLKsZf0" width="533" height="300" frameborder="0" allowfullscreen="" data-external="1"></iframe>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="violin-plots">Violin plots<a class="anchor" aria-label="anchor" href="#violin-plots"></a>
</h3>
<p>The violin plot can show all possible combinations if “show
Permutations” is selected, otherwise a normal violoin plot is shown (not
shown here).</p>
<div class="figure" style="text-align: center">
<img src="images/violinPermutation.png" alt="Violin plot with permutataions" width="90%"><p class="caption">
Violin plot with permutataions
</p>
</div>
<p>Under “grouped” a second variable can be used. Here, grp1-7 are
clusters of cells that have been identified from the UMAP
presentation.</p>
<div class="figure" style="text-align: center">
<img src="images/groupedViolin.png" alt="grouped violin plots" width="90%"><p class="caption">
grouped violin plots
</p>
</div>
</div>
<div class="section level3">
<h3 id="three-dimensional-plot">Three dimensional plot<a class="anchor" aria-label="anchor" href="#three-dimensional-plot"></a>
</h3>
<p>The sum of expression for a list of genes is shown in the 3D tSNE
plot that can be found under Data Exploration -</p>
<div class="figure" style="text-align: center">
<img src="images/3dDEExpression.png" alt="3D tSNE plot with expression of specified genes" width="90%"><p class="caption">
3D tSNE plot with expression of specified genes
</p>
</div>
<div class="figure" style="text-align: center">
<img src="images/scran_17.png" alt="Panel Plot" width="90%"><p class="caption">
Panel Plot
</p>
</div>
</div>
<div class="section level3">
<h3 id="differential-gene-expression-dge">Differential gene expression (DGE)<a class="anchor" aria-label="anchor" href="#differential-gene-expression-dge"></a>
</h3>
<p>This short video shows how to perform a DGE analysis. The individual
steps are also explained afterwards in more detail.</p>
<div class="vembedr">
<div>
<iframe src="https://www.youtube.com/embed/lFHfZKt3zro" width="533" height="300" frameborder="0" allowfullscreen="" data-external="1"></iframe>
</div>
</div>
<div class="section level4">
<h4 id="selections-of-groups-of-cells">selections of groups of cells<a class="anchor" aria-label="anchor" href="#selections-of-groups-of-cells"></a>
</h4>
<p>The “selection of input cells” allows subsetting the cells even
before the visualization.</p>
<p>Cells to be compared can be either selected in the two plot directly,
of if cells are only selected in the first plot, all other cells are
being used.</p>
<div class="figure" style="text-align: center">
<img src="images/scran_18.png" alt="DGE between specific groups of cells" width="90%"><p class="caption">
DGE between specific groups of cells
</p>
</div>
</div>
<div class="section level4">
<h4 id="volcano-plot">Volcano plot<a class="anchor" aria-label="anchor" href="#volcano-plot"></a>
</h4>
<p>Genes can be selected in the volcano plot and their names/symbols are
shown above the plot to be copy/pasted.</p>
<div class="figure" style="text-align: center">
<img src="images/scran_19.png" alt="DGE results as Volcano plot" width="90%"><p class="caption">
DGE results as Volcano plot
</p>
</div>
</div>
<div class="section level4">
<h4 id="dge-table">DGE table<a class="anchor" aria-label="anchor" href="#dge-table"></a>
</h4>
<p>A complete table is available, that can also be downloaded.</p>
<div class="figure" style="text-align: center">
<img src="images/scran_20.png" alt="Table of DGE" width="90%"><p class="caption">
Table of DGE
</p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="comparing-different-clustering-methods">Comparing different clustering methods<a class="anchor" aria-label="anchor" href="#comparing-different-clustering-methods"></a>
</h3>
<p>SCHNAPPs uses the projection “dbCluster” for the current clustering
result. In order to compare different clustering procedures the rename
projection tool allows the user to copy the results do a different
name.</p>
<div class="figure" style="text-align: center">
<img src="images/scran_21.png" alt="Projections can be renamed" width="90%"><p class="caption">
Projections can be renamed
</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="please-help">Please help<a class="anchor" aria-label="anchor" href="#please-help"></a>
</h2>
<p>We hope you find SCHNAPPs useful. Please don’t forget to cite our
work when you publish, nor tell us if you find any bugs or have
questions/suggestions.</p>
<p>bernd <dot> jagla <at> pasteur <dot> fr</dot></at></dot></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Bernd Jagla.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
