<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Seurat workflow • SCHNAPPs</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><meta property="og:title" content="Seurat workflow">
<meta property="og:description" content="SCHNAPPs">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">SCHNAPPs</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.4.61</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/Readme.html">Readme</a>
    </li>
    <li>
      <a href="../../articles/pkdown/FAQs.html">Frequently asked questions for common tasks</a>
    </li>
    <li>
      <a href="../../articles/pkdown/SCHNAPPs_usage.html">SCHNAPPs - Single Cell sHiNy APP(s)</a>
    </li>
    <li>
      <a href="../../articles/pkdown/SeuratWorkflow.html">Seurat workflow</a>
    </li>
    <li>
      <a href="../../articles/pkdown/scranWorkflow.html">Follow scran workflow to analyze single-cell RNA-seq data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/C3BI-pasteur-fr/UTechSCB-SCHNAPPs/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="SeuratWorkflow_files/header-attrs-2.2/header-attrs.js"></script><script src="SeuratWorkflow_files/htmlwidgets-1.5.1/htmlwidgets.js"></script><script src="SeuratWorkflow_files/jquery-1.12.4/jquery.min.js"></script><link href="SeuratWorkflow_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="SeuratWorkflow_files/datatables-binding-0.13/datatables.js"></script><link href="SeuratWorkflow_files/dt-core-1.10.20/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="SeuratWorkflow_files/dt-core-1.10.20/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="SeuratWorkflow_files/dt-core-1.10.20/js/jquery.dataTables.min.js"></script><link href="SeuratWorkflow_files/crosstalk-1.1.0.1/css/crosstalk.css" rel="stylesheet">
<script src="SeuratWorkflow_files/crosstalk-1.1.0.1/js/crosstalk.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Seurat workflow</h1>
                        <h4 class="author">Bernd Jagla</h4>
                  <a class="author_email" href="mailto:#">bernd <dot> jagla <at> pasteur <dot> fr</dot></at></dot></a>
      
                  
            <h4 class="date">6/7/2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/C3BI-pasteur-fr/UTechSCB-SCHNAPPs/blob/master/vignettes/pkdown/SeuratWorkflow.Rmd"><code>vignettes/pkdown/SeuratWorkflow.Rmd</code></a></small>
      <div class="hidden name"><code>SeuratWorkflow.Rmd</code></div>

    </div>

    
    
<p>Please see the scran workflow for a more indepth workflow.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">dplyr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">Seurat</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">patchwork</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">SCHNAPPs</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">SingleCellExperiment</span>)
<span class="co"># Load the pbm dataset</span>
<span class="no">pbm.data</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/Read10X.html">Read10X</a></span>(<span class="kw">data.dir</span> <span class="kw">=</span> <span class="st">"~/Downloads/filtered_gene_bc_matrices/hg19/"</span>)
<span class="co"># Initialize the Seurat object with the raw (non-normalized data).</span>
<span class="no">pbm</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/CreateSeuratObject.html">CreateSeuratObject</a></span>(<span class="kw">counts</span> <span class="kw">=</span> <span class="no">pbm.data</span>, <span class="kw">project</span> <span class="kw">=</span> <span class="st">"pbm3k"</span>, <span class="kw">min.cells</span> <span class="kw">=</span> <span class="fl">3</span>, <span class="kw">min.features</span> <span class="kw">=</span> <span class="fl">200</span>)

<span class="no">pbm</span><span class="kw">[[</span><span class="st">"percent.mt"</span>]] <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/PercentageFeatureSet.html">PercentageFeatureSet</a></span>(<span class="no">pbm</span>, <span class="kw">pattern</span> <span class="kw">=</span> <span class="st">"^MT-"</span>)

<span class="co">## in case of human samples:</span>
<span class="co"># pbm &lt;- CellCycleScoring(</span>
<span class="co"># pbm,</span>
<span class="co">#   g2m.features = cc.genes$g2m.genes,</span>
<span class="co">#   s.features = cc.genes$s.genes</span>
<span class="co"># )</span>



<span class="no">scEx</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/as.SingleCellExperiment.html">as.SingleCellExperiment</a></span>(<span class="no">pbm</span>)

<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="fu">colData</span>(<span class="no">scEx</span>)) <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"sampleNames"</span>,   <span class="st">"nCount_RNA"</span>,   <span class="st">"nFeature_RNA"</span>, <span class="st">"percent.mt"</span> , <span class="st">"ident"</span>       )
<span class="fu">colData</span>(<span class="no">scEx</span>)$<span class="no">barcode</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="fu">colData</span>(<span class="no">scEx</span>))

<span class="fu">rowData</span>(<span class="no">scEx</span>)$<span class="no">Description</span> <span class="kw">=</span> <span class="st">""</span>
<span class="fu">rowData</span>(<span class="no">scEx</span>)$<span class="no">id</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="fu">rowData</span>(<span class="no">scEx</span>))
<span class="fu">rowData</span>(<span class="no">scEx</span>)$<span class="no">symbol</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="fu">rowData</span>(<span class="no">scEx</span>))

<span class="no">pbm</span></pre></body></html></div>
<pre><code>## An object of class Seurat 
## 13714 features across 2700 samples within 1 assay 
## Active assay: RNA (13714 features, 0 variable features)</code></pre>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="no">pbm</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/subset.Seurat.html">subset</a></span>(<span class="no">pbm</span>, <span class="kw">subset</span> <span class="kw">=</span> <span class="no">nFeature_RNA</span> <span class="kw">&gt;</span> <span class="fl">200</span> <span class="kw">&amp;</span> <span class="no">nFeature_RNA</span> <span class="kw">&lt;</span> <span class="fl">2500</span> <span class="kw">&amp;</span> <span class="no">percent.mt</span> <span class="kw">&lt;</span> <span class="fl">5</span>)
<span class="no">pbm</span> <span class="kw">&lt;-</span> <span class="no">pbm</span>[-<span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span>(<span class="no">pbm</span>)<span class="kw">==</span><span class="fl">0</span>),]
<span class="no">pbm</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/NormalizeData.html">NormalizeData</a></span>(<span class="no">pbm</span>)

<span class="no">pbm</span> <span class="kw">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindVariableFeatures.html">FindVariableFeatures</a></span>(<span class="no">pbm</span>, <span class="kw">selection.method</span> <span class="kw">=</span> <span class="st">"vst"</span>, <span class="kw">nfeatures</span> <span class="kw">=</span> <span class="fl">2000</span>)

<span class="no">top10</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/VariableFeatures.html">VariableFeatures</a></span>(<span class="no">pbm</span>), <span class="fl">10</span>)
<span class="no">all.VariableFeatures</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/VariableFeatures.html">VariableFeatures</a></span>(<span class="no">pbm</span>)
<span class="no">all.genes</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">pbm</span>)
<span class="co"># Why do they scale on all features???</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">1</span>)
<span class="no">pbm</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/ScaleData.html">ScaleData</a></span>(<span class="no">pbm</span>, <span class="kw">features</span> <span class="kw">=</span> <span class="no">all.VariableFeatures</span>)

<span class="no">pbm</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/RunPCA.html">RunPCA</a></span>(<span class="no">pbm</span>, <span class="kw">features</span> <span class="kw">=</span> <span class="no">all.VariableFeatures</span>, <span class="kw">rank</span> <span class="kw">=</span> <span class="fl">50</span>)
<span class="no">scranPCA</span> <span class="kw">=</span> <span class="kw pkg">BiocSingular</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html">runPCA</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>((<span class="kw pkg">Seurat</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/Assays.html">Assays</a></span>(<span class="no">pbm</span>,<span class="kw">slot</span> <span class="kw">=</span> <span class="st">"RNA"</span>)@<span class="kw">scale.data</span>))[<span class="no">all.VariableFeatures</span>,]) , <span class="kw">rank</span> <span class="kw">=</span> <span class="fl">50</span>)

<span class="no">pbm</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindNeighbors.html">FindNeighbors</a></span>(<span class="no">pbm</span>, <span class="kw">dims</span> <span class="kw">=</span> <span class="fl">1</span>:<span class="fl">10</span>)
<span class="no">pbm</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindClusters.html">FindClusters</a></span>(<span class="no">pbm</span>, <span class="kw">resolution</span> <span class="kw">=</span> <span class="fl">0.5</span>, <span class="kw">verbose</span> <span class="kw">=</span> <span class="fl">FALSE</span>)

<span class="no">pbm</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/RunUMAP.html">RunUMAP</a></span>(<span class="no">pbm</span>, <span class="kw">dims</span> <span class="kw">=</span> <span class="fl">1</span>:<span class="fl">10</span>, <span class="kw">verbose</span> <span class="kw">=</span> <span class="fl">FALSE</span>)

<span class="fu">colData</span>(<span class="no">scEx</span>)$<span class="no">seurartCluster</span> <span class="kw">=</span> -<span class="fl">1</span>
<span class="fu">colData</span>(<span class="no">scEx</span>)[<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/Idents.html">Idents</a></span>(<span class="no">pbm</span>)),<span class="st">"seurartCluster"</span>] <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/Idents.html">Idents</a></span>(<span class="no">pbm</span>)
<span class="fu">colData</span>(<span class="no">scEx</span>)$<span class="no">seurartCluster</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span>(<span class="fu">colData</span>(<span class="no">scEx</span>)$<span class="no">seurartCluster</span>)</pre></body></html></div>
<div id="comparing-biocsingular-and-seurat-pca-calculations" class="section level2">
<h2 class="hasAnchor">
<a href="#comparing-biocsingular-and-seurat-pca-calculations" class="anchor"></a>comparing BiocSingular and Seurat PCA calculations:</h2>
<ul>
<li>
<p>Seurat </p>
<div id="htmlwidget-7af8561a869674bbc463" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-7af8561a869674bbc463">{"x":{"filter":"none","data":[["AAACATACAACCAC-1","AAACATTGAGCTAC-1","AAACATTGATCAGC-1"],[-4.7296855122495,-0.517402927207163,-3.1891063418811],[-0.518426513217326,4.59189565771113,-3.46951535776337],[-0.76232200534983,5.90919209309302,-0.831371037862025]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>PC_1<\/th>\n      <th>PC_2<\/th>\n      <th>PC_3<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":[1,2,3]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</li>
<li>
<p>BiocSingular </p>
<div id="htmlwidget-224a177bc8d0026d195b" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-224a177bc8d0026d195b">{"x":{"filter":"none","data":[["AAACATACAACCAC-1","AAACATTGAGCTAC-1","AAACATTGATCAGC-1"],[-4.71606885483825,-0.503934612562218,-3.17556831366895],[-0.526504991038634,4.58419850565029,-3.47741474467558],[-0.848004823269263,5.88435595219519,-0.912638100491363]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>PC1<\/th>\n      <th>PC2<\/th>\n      <th>PC3<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":[1,2,3]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</li>
</ul>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="no">pbm</span></pre></body></html></div>
<pre><code>## An object of class Seurat 
## 13713 features across 2638 samples within 1 assay 
## Active assay: RNA (13713 features, 2000 variable features)
##  2 dimensional reductions calculated: pca, umap</code></pre>
<div class="sourceCode" id="cb6"><html><body><pre class="r"># Assays(pbm)</pre></body></html></div>
<p>We have now created a SingleCellExperiment object with all information from Seurat. We will now try to recreate these results with SCHNAPPs:</p>
<p>We have to save the object in a file that can be opened with the “load” command.</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/save.html">save</a></span>(<span class="kw">file</span> <span class="kw">=</span> <span class="st">"seurat.pbm.RData"</span>, <span class="kw">list</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"scEx"</span>))</pre></body></html></div>
<p>To reproduce the results the following parameters have to be set in SCHNAPPs:</p>
<ul>
<li>Cell selection: ** Min # of UMIs = 1</li>
</ul>
<div class="figure" style="text-align: center">
<img src="images/scran_1a.png" alt="Cell selection parameters" width="70%"><p class="caption">
Cell selection parameters
</p>
</div>
<p>Just to be complete (we will do this later) here is the list of cells that has to be removed:</p>
<p>AAAGATCTGGGCAA-1, AAAGCAGAAGCCAT-1, AACGCCCTGCTTAG-1, AAGGTCTGGTATGC-1, AATGTAACGTTTGG-1, AATTACGAGTAGCT-1, ACACAGACACCTGA-1, ACATGGTGCGTTGA-1, ACCTGGCTGTCTTT-1, ACTTAAGACCACAA-1, ACTTGTACCCGAAT-1, ACTTTGTGCGATAC-1, AGAGGTCTACAGCT-1, ATCACGGATTGCTT-1, ATTACCTGGGCATT-1, CACGCTACTTGACG-1, CAGTGTGAACACGT-1, CCAATGGAACAGCT-1, CCAGTCTGCGGAGA-1, CGACCTTGGCAAGG-1, CGAGCCGACGACAT-1, CGGAATTGCACTAG-1, CGTAACGAATCAGC-1, CGTACCACGCTACA-1, CGTACCTGGACGAG-1, CTAGTTTGAGTACC-1, CTCAGCTGTTTCTG-1, CTCATTGATTGCTT-1, CTGGCACTGGACAG-1, CTTAACACGAGCTT-1, CTTAAGCTTCCTCG-1, GAAAGATGTTTGCT-1, GAACGTTGACGGAG-1, GAATGGCTAAGATG-1, GACCATGACTCTCG-1, GACTGAACAACCGT-1, GCCACTACCTACTT-1, GCGAAGGAGAGCTT-1, GCTACAGATCTTAC-1, GGCACGTGTGAGAA-1, GTCAACGATCAGGT-1, GTGAACACAGATCC-1, GTGTCAGAATGCTG-1, GTTAAAACTTCGCC-1, TAAGATACCCACAA-1, TACGCAGACGTCTC-1, TACGCGCTCTTCTA-1, TACGGCCTGTCCTC-1, TATCACTGACTGTG-1, TCCCGATGCTGTGA-1, TCGCACACCATCAG-1, TCGTGAGAACTGTG-1, TGAAGCTGAGACTC-1, TGAGACACTGTGCA-1, TGAGCTGAGCGAGA-1, TGGAGACTGAAACA-1, TGGATGTGATGTCG-1, TGGCAATGGAGGGT-1, TGGTCAGACCGTTC-1, TGTTAAGATTGGCA-1, TTACTCGAACGTTG-1, TTCAAGCTTCCAAG-1</p>
<div class="figure" style="text-align: center">
<img src="images/cellAddParam.png" alt="Cell selection additional parameters" width="70%"><p class="caption">
Cell selection additional parameters
</p>
</div>
<ul>
<li>Gene selection ** remove regular expression ** set min expression to 1</li>
</ul>
<div class="figure" style="text-align: center">
<img src="images/geneSelection.png" alt="Gene selection parameters" width="70%"><p class="caption">
Gene selection parameters
</p>
</div>
<ul>
<li>Parameters - Normalization</li>
</ul>
<p>** Select SeuratLogNorm, the standard Normalization function in Seurat.</p>
<div class="figure" style="text-align: center">
<img src="images/SeuratNormalization.png" alt="Select normalization method" width="70%"><p class="caption">
Select normalization method
</p>
</div>
<ul>
<li><p>Parameters - General Parameters</p></li>
<li><p>PCA parameters</p></li>
</ul>
<p>** center = true ** scate = true ** use Seurat::RunPCA = ture ** Number of variable genes to be used = 2000 ** select highly expressed using “vst”</p>
<ul>
<li>Clustering parameters ** Seurat clustering tab. ** Dimensions from PCA to use: 10 ** k = 20</li>
</ul>
<div class="figure" style="text-align: center">
<img src="images/seuratGeneralParemeters.png" alt="General parameters for PCA and clustering" width="70%"><p class="caption">
General parameters for PCA and clustering
</p>
</div>
<ul>
<li>load input file</li>
</ul>
<p>** unset sub sampling of data ** check “calculate normalization here”</p>
<div class="figure" style="text-align: center">
<img src="images/inputOptions.png" alt="Load data - parameters" width="70%"><p class="caption">
Load data - parameters
</p>
</div>
<div class="figure" style="text-align: center">
<img src="images/loadSeurat.png" alt="Load data select file" width="70%"><p class="caption">
Load data select file
</p>
</div>
</div>
<div id="set-color-for-sample" class="section level2">
<h2 class="hasAnchor">
<a href="#set-color-for-sample" class="anchor"></a>Set color for sample</h2>
<p>Just for estetics:</p>
<p>Under Parameters - General Parameters set a nicer color for the samples other than black (which isn’t a color anyways.)</p>
<div class="figure" style="text-align: center">
<img src="images/setColor.png" alt="Set color for samples" width="70%"><p class="caption">
Set color for samples
</p>
</div>
</div>
<div id="select-cells-to-be-removed" class="section level2">
<h2 class="hasAnchor">
<a href="#select-cells-to-be-removed" class="anchor"></a>select cells to be removed</h2>
<p>Using the 2D plot the cells with more than 2500 nFeature_RNA and more than 5 percent.mt can be selected and removed from the data set:</p>
<p>We have to comply to three thresholds that are being used:</p>
<ul>
<li>nFeature_RNA &gt; 200</li>
<li>nFeature_RNA &lt; 2500</li>
<li>percent.mt &lt; 5</li>
</ul>
<p>Go to Co-expression - selected:</p>
<p>set the following options:</p>
<ul>
<li>X : nFeature_RNA</li>
<li>Y : percent.mt</li>
<li>color: sampleNames</li>
</ul>
<div class="figure" style="text-align: center">
<img src="images/selectionSeurat.png" alt="2D plot, select cells to be removed 1" width="70%"><p class="caption">
2D plot, select cells to be removed 1
</p>
</div>
<p>select the zoom function in the plot:</p>
<div class="figure" style="text-align: center">
<img src="images/zoom.png" alt="zooming selector" width="30%"><p class="caption">
zooming selector
</p>
</div>
<p>Zooming into the low nFeatureRNA region reveals that no cells will be removed using the 200 threshold.</p>
<div class="figure" style="text-align: center">
<img src="images/2Dzoomed.png" alt="Filter based on detected genes" width="70%"><p class="caption">
Filter based on detected genes
</p>
</div>
<p>choose box select:</p>
<div class="figure" style="text-align: center">
<img src="images/boxSelect.png" alt="Selection tool (box)" width="30%"><p class="caption">
Selection tool (box)
</p>
</div>
<ul>
<li>Select the cells that are higher than 5 (roughly) percent.mt</li>
<li>set the name of the group to “rmCells”</li>
<li>group names has to be plot</li>
</ul>
<div class="figure" style="text-align: center">
<img src="images/selecteHighMT.png" alt="2D plot, select cells to be removed 2" width="70%"><p class="caption">
2D plot, select cells to be removed 2
</p>
</div>
<ul>
<li>click on current selection</li>
</ul>
<p>=&gt; the cells are now red</p>
<ul>
<li>set group names back to “plot”</li>
</ul>
<p>change the zoom to be able to select anything around 5 percent.mt and show additional options:</p>
<div class="figure" style="text-align: center">
<img src="images/zoomedSelection.png" alt="2D plot, select cells to be removed 3" width="70%"><p class="caption">
2D plot, select cells to be removed 3
</p>
</div>
<ul>
<li>select the remaining cells and add them to “rmCells” list.</li>
<li>if you are not sure about a cell the values can be shown hovering over the points</li>
</ul>
<div class="figure" style="text-align: center">
<img src="images/indivCells.png" alt="2D plot, select cells to be removed 4" width="70%"><p class="caption">
2D plot, select cells to be removed 4
</p>
</div>
<ul>
<li>proceed similarly to select the cells that have more than 2500 nFeature_RNA.</li>
</ul>
<p>You should have 62 cells selected.</p>
<ul>
<li>check the “show cell names” check box.</li>
<li>copy the cell names (tripple click on a cell name; Command-C)</li>
<li>Paste the cell name under “Cells to be removed” (see above, cell selection)</li>
<li>click “apply changes”</li>
</ul>
<div class="figure" style="text-align: center">
<img src="images/showNames.png" alt="2D plot, select cells to be removed 4" width="70%"><p class="caption">
2D plot, select cells to be removed 4
</p>
</div>
<ul>
<li>summary Stats:</li>
</ul>
<p>That is how summary stats should look like:</p>
<div class="figure" style="text-align: center">
<img src="images/summaryStatsSeurat.png" alt="Summary statistics" width="50%"><p class="caption">
Summary statistics
</p>
</div>
</div>
<div id="clustering-comparison" class="section level2">
<h2 class="hasAnchor">
<a href="#clustering-comparison" class="anchor"></a>Clustering comparison</h2>
<p>We can now compare the clustering results:</p>
<p>Both graphs show the same thing: All cells have been assigned to the same cluster. (Just the numbering is different.)</p>
<div class="figure" style="text-align: center">
<img src="images/alluvialSeurat.png" alt="Alluvial plot" width="90%"><p class="caption">
Alluvial plot
</p>
</div>
<div class="figure" style="text-align: center">
<img src="images/2dSeuratCluster.png" alt="2D cluster comparison" width="90%"><p class="caption">
2D cluster comparison
</p>
</div>
<p>Q.E.D.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Bernd Jagla.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
